<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Ariane L. Moore et al." />

<meta name="date" content="2018-04-20" />

<title>Package GeneAccord</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Package ‘GeneAccord’</h1>
<h4 class="author"><em>Ariane L. Moore et al.</em></h4>
<h4 class="date"><em>2018-04-20</em></h4>



<div id="detection-of-clonally-exclusive-gene-or-pathway-pairs-in-a-cohort-of-cancer-patients" class="section level1">
<h1>Detection of clonally exclusive gene or pathway pairs in a cohort of cancer patients</h1>
<p>GeneAccord is a statistical framework to examine the combinations of clones that co-exist in tumors [1]. More precisely, the algorithm finds pairs of genes that are mutated in the same tumor but in different clones, i.e. their subclonal mutation profiles are mutually exclusive. We refer to this as clonally exclusive. It means that the mutations occurred in different branches of the tumor phylogeny. The motivation to detect such gene pairs is to uncover potential mechanisms of synergies between co-existing clones in tumors. Our statistical framework assesses whether a pattern of clonal exclusivity occurs more often than expected by chance alone across a cohort of patients. It makes use of a likelihood ratio test to compare the number of observed clonal exclusivities of a gene pair of interest across the cohort to the background distribution of the expected frequencies of clonal exclusivity. The null hypothesis is that the gene pair of interest exhibits the pattern of clonal exclusivity as often as expected by chance when assuming independent occurrence of mutations. The alternative hypothesis is that the pattern occurs at a rate greater than random chance. Multiple testing correction is done using the Benjamini-Hochberg procedure [<a href="https://pdfs.semanticscholar.org/af6e/9cd1652b40e219b45402313ec6f4b5b3d96b.pdf">2</a>]. If a pair is significantly clonally exclusive, it suggests that this specific clone configuration confers a selective advantage, possibly through synergies between the clones with these mutations. The required input data are the mutated gene-to-clone assignments from a cohort of cancer patients, which were obtained by running phylogenetic tree inference methods. Reconstructing the evolutionary history of a tumor and detecting the clones is challenging [<a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4265145/">3</a>]. For nondeterministic algorithms, repeated tree inference runs may lead to slightly different mutation-to-clone assignments. Therefore, our algorithm was designed to allow the input of multiple gene-to-clone assignments per patient. They may have been generated by repeatedly performing the tree inference, or by sampling from the posterior distribution of trees. The tree inference methods designate the mutations to individual clones. The mutations can then be mapped to genes or pathways. Hence our statistical framework can be applied on the gene level, or on the pathway level to detect clonally exclusive pairs of pathways. If you use GeneAccord in your published work, please cite the paper in which it was introduced [1], where you can also find a more detailed description of the statistical framework.</p>
<hr />
<div id="installation" class="section level2">
<h2>Installation</h2>
<p>You can install the package as described in the <code>README</code> of this package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(GeneAccord)</code></pre></div>
<hr />
</div>
<div id="prepare-patient-data" class="section level2">
<h2>1. Prepare patient data</h2>
<p>The first step consists of reading in and processing the patient data separately for each patient. If it takes too much time, this can be done for all patients in parallel on a cluster. In this example, we read in the data for patient “01”. The example data is stored in the external data directory of this package. The phylogenetic tree inference was performed with the R package <code>Cloe</code> [<a href="https://doi.org/10.1214/16-AOAS986">4</a>], which was run 20 times with 20 different seeds (seed = 5, 10, 15, …, 100) to capture the uncertainty in the phylogenetic tree inference. All 20 gene-to-clone assignments for patient 01 can be found as follows.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># defines the path to the external data directory</span>
ext_data_dir &lt;-<span class="st"> </span><span class="kw">system.file</span>(<span class="st">'extdata'</span>, <span class="dt">package =</span> <span class="st">'GeneAccord'</span>)

<span class="co"># set the patient id</span>
pat_id &lt;-<span class="st"> &quot;01&quot;</span>

<span class="co"># find all csv-files from patient 01, which are in the clonal_genotypes subdirectory, with seeds 5, 10, 15, ..., 100</span>
input_files_<span class="dv">01</span> &lt;-<span class="st"> </span><span class="kw">paste</span>(ext_data_dir, <span class="st">&quot;/clonal_genotypes/cloe_seed&quot;</span>, <span class="kw">seq</span>(<span class="dv">5</span>, <span class="dv">100</span>, <span class="dt">by =</span> <span class="dv">5</span>), <span class="st">&quot;/&quot;</span>, pat_id, <span class="st">&quot;.csv&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;&quot;</span>)
<span class="co"># 'input_files_01' is now a vector with all 20 csv-files for this patient</span></code></pre></div>
<p>So to analyze your own data, you need to set a vector with all gene-to-clone assignments in csv-format for each patient. Such a csv-file may look like this.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(<span class="kw">read.csv</span>(input_files_<span class="dv">01</span>[<span class="dv">1</span>]))
<span class="co">#&gt;    mutated_entity clone1 clone2 clone3 clone4</span>
<span class="co">#&gt; 1 ENSG00000008197      0      1      1      1</span>
<span class="co">#&gt; 2 ENSG00000051128      0      1      1      1</span>
<span class="co">#&gt; 3 ENSG00000078403      0      0      1      1</span>
<span class="co">#&gt; 4 ENSG00000084774      0      1      1      1</span>
<span class="co">#&gt; 5 ENSG00000100196      0      1      1      1</span>
<span class="co">#&gt; 6 ENSG00000105976      0      0      1      1</span></code></pre></div>
<p>The first column lists the altered genes or pathways. In this manual, we will use the term genes, even if the altered entities can also be pathways. The columns that come after the gene indicate in which clone the gene is mutated. In this package, data tables or csv-files are stored and used as tibbles. Tibbles are a novel way to represent data frames. Please refer to the R package <code>tibble</code> [<a href="https://CRAN.R-project.org/package=tibble">5</a>] for more information. Now we read in the patient’s gene-to-clone assignments from the collection of tree inferences to obtain it in a tibble format.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clone_tbl_01_all_trees &lt;-<span class="st"> </span><span class="kw">create_tbl_tree_collection</span>(input_files_<span class="dv">01</span>)
<span class="co">#&gt; Found a collection of tree inferences of 20 input files.</span>
<span class="co">#&gt; Removed 1 gene that is missing in more than 2 of gene-to-clone assignments.</span>
<span class="co">#&gt; Found 580 altered genes/pathways in total, which are from 20 gene-to-clone assignments.</span></code></pre></div>
<p>Let’s look at the tibble:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clone_tbl_01_all_trees
<span class="co">#&gt; # A tibble: 580 x 11</span>
<span class="co">#&gt;    file_name  patient_id altered_entity clone1 clone2 clone3 clone4 clone5</span>
<span class="co">#&gt;    &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;           &lt;int&gt;  &lt;int&gt;  &lt;int&gt;  &lt;int&gt;  &lt;int&gt;</span>
<span class="co">#&gt;  1 cloe_seed~ 01         ENSG000000081~      0      1      1      1      0</span>
<span class="co">#&gt;  2 cloe_seed~ 01         ENSG000000511~      0      1      1      1      0</span>
<span class="co">#&gt;  3 cloe_seed~ 01         ENSG000000784~      0      0      1      1      0</span>
<span class="co">#&gt;  4 cloe_seed~ 01         ENSG000000847~      0      1      1      1      0</span>
<span class="co">#&gt;  5 cloe_seed~ 01         ENSG000001001~      0      1      1      1      0</span>
<span class="co">#&gt;  6 cloe_seed~ 01         ENSG000001059~      0      0      1      1      0</span>
<span class="co">#&gt;  7 cloe_seed~ 01         ENSG000001082~      0      0      1      1      0</span>
<span class="co">#&gt;  8 cloe_seed~ 01         ENSG000001088~      0      1      1      1      0</span>
<span class="co">#&gt;  9 cloe_seed~ 01         ENSG000001112~      0      0      1      1      0</span>
<span class="co">#&gt; 10 cloe_seed~ 01         ENSG000001137~      0      1      1      1      0</span>
<span class="co">#&gt; # ... with 570 more rows, and 3 more variables: clone6 &lt;int&gt;,</span>
<span class="co">#&gt; #   clone7 &lt;int&gt;, tree_id &lt;int&gt;</span></code></pre></div>
<p>The column <code>file_name</code> indicates from which csv-file the information comes. The patient identifier is stored in <code>patient_id</code>, and the column <code>altered_entity</code> lists the mutated gene. The gene-to-clone assignments are in the columns <code>clone1</code>, <code>clone2</code>, …, up to, by default <code>clone7</code>. Even if this patient had fewer clones in the assignment, the additional clone columns make sure that all data from all patients can be stored in the same tibble. The additional clone columns are just always zero, if a patient had fewer clones. If you know that you have more than seven clones at maximum across all patients and trees, the parameter <code>max_num_clones</code> of <code>create_tbl_tree_collection</code> can be set accordingly. The last column in the tibble <code>tree_id</code> indicates from which of the trees the respective gene-to-clone assignment comes. In this tibble, we have 20 different tree identifiers, because we use the gene-to-clone assignments from 20 tree inference runs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># print all 20 tree id's</span>
<span class="kw">unique</span>(<span class="kw">as.character</span>(clone_tbl_01_all_trees<span class="op">$</span>tree_id))
<span class="co">#&gt;  [1] &quot;1&quot;  &quot;2&quot;  &quot;3&quot;  &quot;4&quot;  &quot;5&quot;  &quot;6&quot;  &quot;7&quot;  &quot;8&quot;  &quot;9&quot;  &quot;10&quot; &quot;11&quot; &quot;12&quot; &quot;13&quot; &quot;14&quot;</span>
<span class="co">#&gt; [15] &quot;15&quot; &quot;16&quot; &quot;17&quot; &quot;18&quot; &quot;19&quot; &quot;20&quot;</span></code></pre></div>
<p>The next step is to compute the rate of clonal exclusivity for this patient, and for each of the tree inferences separately.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rates_clon_excl_<span class="dv">01</span> &lt;-<span class="st"> </span><span class="kw">compute_rates_clon_excl</span>(clone_tbl_01_all_trees)
<span class="co">#&gt; Found 20 different tree instances in the gene-to-clone tibble from patient 01.</span>
<span class="co">#&gt; Will compute the rate of clonal exclusivity for each of these...</span>
<span class="co">#&gt; ...done</span>
<span class="co">#&gt; Current patient 01 has mean rate 0.006, median rate 0, and the rate is zero in 16 out of 20 trees.</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(rates_clon_excl_<span class="dv">01</span>)
<span class="co">#&gt;         01         01         01         01         01         01 </span>
<span class="co">#&gt; 0.03448276 0.00000000 0.01477833 0.03448276 0.03448276 0.00000000</span>

<span class="co"># average rate for patient 01 across trees</span>
m_<span class="dv">01</span> &lt;-<span class="st"> </span><span class="kw">mean</span>(rates_clon_excl_<span class="dv">01</span>)</code></pre></div>
<p>Now, for each gene pair separately, two values are computed: The number of times the gene pair occurs in the collection of trees, and the number of times it was clonally exclusive in these trees. A gene pair does not have to occur in all trees, because sometimes a gene is not assigned to a clone in a tree. Hence, the number of times a pair occurs may be, e.g. 19 of 20 times. And the number of times a pair is clonally exlcusive in the trees of course varies from gene pair to gene pair.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># compute the histogram of how often gene pairs occur in the trees, and how often they are clonally exclusive</span>
hist_of_num_trees_clon_excl_<span class="dv">01</span> &lt;-<span class="st"> </span><span class="kw">get_hist_clon_excl</span>(clone_tbl_01_all_trees)
<span class="co">#&gt; Found 20 different tree instances in the gene-to-clone tibble from patient 01, and 29 different mutated genes/pathways in total.</span>
<span class="co">#&gt; Will compute the histograms for all pairs of how often they occur across tree instances, and how often they are clonally exclusive...</span>
<span class="co">#&gt; ...done</span>
<span class="co">#&gt; The median value for how often a gene pair occurs across the trees is 20, and 0 is the median value for the pairs being clonally exclusive.</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(hist_of_num_trees_clon_excl_<span class="dv">01</span>[[<span class="dv">1</span>]])
<span class="co">#&gt; [1] 20 20 20 20 20 20</span>
<span class="kw">head</span>(hist_of_num_trees_clon_excl_<span class="dv">01</span>[[<span class="dv">2</span>]])
<span class="co">#&gt; [1] 0 0 0 0 0 0</span></code></pre></div>
<p>We observe, for instance, that some gene pairs occur in 20 trees, and are always co-occurring in the same clones.</p>
<p>The three steps demonstrated here (<code>create_tbl_tree_collection</code>, <code>compute_rates_clon_excl</code>, and <code>get_hist_clon_excl</code>) for patient 01 have to be done for all patients in the cohort. If you have a high number of patients, you can run it in parallel on a computing cluster. If you choose to run it sequentially, the following chunk of code is an example of how this would need to be done.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># set the patient identifiers</span>
all_pat_ids &lt;-<span class="st"> </span><span class="kw">c</span>()
<span class="cf">for</span>(i <span class="cf">in</span> <span class="kw">seq</span>(<span class="dv">1</span>, <span class="dv">89</span>)){
  <span class="cf">if</span>(i <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="dv">3</span>, <span class="dv">34</span>, <span class="dv">51</span>, <span class="dv">53</span>, <span class="dv">58</span>, <span class="dv">65</span>, <span class="dv">87</span>)) {
    <span class="cf">next</span>
  } 
  all_pat_ids &lt;-<span class="st"> </span><span class="kw">c</span>(all_pat_ids, <span class="kw">ifelse</span>(i<span class="op">&lt;</span><span class="dv">10</span>, <span class="kw">paste0</span>(<span class="st">&quot;0&quot;</span>, i), <span class="kw">as.character</span>(i)))
}

<span class="co"># this will be the vector with the average rates of clonal exclusivity of each patient</span>
avg_rates_m &lt;-<span class="st"> </span><span class="kw">c</span>()

<span class="co"># these will be the two lists with an entry for each patient with</span>
<span class="co"># a histogram of how often pairs occur in the collection of trees</span>
list_of_num_trees_all_pats &lt;-<span class="st"> </span><span class="kw">list</span>()
<span class="co"># a histogram of how ofthen pairs were clonally exclusive</span>
list_of_clon_excl_all_pats &lt;-<span class="st"> </span><span class="kw">list</span>()
<span class="co"># this is the list of tibbles that will contain the gene-to-clone assignments from all patients from all trees</span>
clone_tbl_all_pats_all_trees_list &lt;-<span class="st"> </span><span class="kw">list</span>()


<span class="co"># loop over all patients and prepare the input data</span>
cnt &lt;-<span class="st"> </span><span class="dv">0</span>
<span class="cf">for</span> (this_pat <span class="cf">in</span> all_pat_ids){
  cnt &lt;-<span class="st"> </span>cnt <span class="op">+</span><span class="st"> </span><span class="dv">1</span>

  <span class="co"># find all csv-files from current patient, which are in the clonal_genotypes subdirectory, with seeds 5, 10, 15, ..., 100</span>
  input_files_this_pat &lt;-<span class="st"> </span><span class="kw">paste0</span>(ext_data_dir, <span class="st">&quot;/clonal_genotypes/cloe_seed&quot;</span>, <span class="kw">seq</span>(<span class="dv">5</span>, <span class="dv">100</span>, <span class="dt">by =</span> <span class="dv">5</span>), <span class="st">&quot;/&quot;</span>, this_pat, <span class="st">&quot;.csv&quot;</span>)

  <span class="co"># here, there should be files from 20 different seeds</span>
  <span class="kw">stopifnot</span>(<span class="kw">length</span>(input_files_this_pat) <span class="op">==</span><span class="st"> </span><span class="dv">20</span>)
  
  <span class="co"># generate the tibble containing the gene-to-clone-assignments</span>
  clone_tbl_this_pat_all_trees &lt;-<span class="st"> </span><span class="kw">create_tbl_tree_collection</span>(input_files_this_pat)

  <span class="co"># compute the rates of clonal exclusivity</span>
  rates_clon_excl_this_pat &lt;-<span class="st"> </span><span class="kw">compute_rates_clon_excl</span>(clone_tbl_this_pat_all_trees)

  <span class="co"># average rate for current patient across trees</span>
  m_this_pat &lt;-<span class="st"> </span><span class="kw">mean</span>(rates_clon_excl_this_pat)

  <span class="co"># compute the histogram of how often gene pairs occur in the trees, and how often they are clonally exclusive</span>
  hist_of_num_trees_clon_excl_this_pat &lt;-<span class="st"> </span><span class="kw">get_hist_clon_excl</span>(clone_tbl_this_pat_all_trees)

  <span class="co"># save this information into named vectors/lists</span>
  avg_rates_m &lt;-<span class="st"> </span><span class="kw">c</span>(avg_rates_m, m_this_pat)
  list_of_num_trees_all_pats[[cnt]] &lt;-<span class="st"> </span>hist_of_num_trees_clon_excl_this_pat[[<span class="dv">1</span>]]
  list_of_clon_excl_all_pats[[cnt]] &lt;-<span class="st"> </span>hist_of_num_trees_clon_excl_this_pat[[<span class="dv">2</span>]]
  <span class="kw">names</span>(avg_rates_m)[cnt] &lt;-<span class="st"> </span><span class="kw">names</span>(list_of_num_trees_all_pats)[cnt] &lt;-<span class="st"> </span><span class="kw">names</span>(list_of_clon_excl_all_pats)[cnt] &lt;-<span class="st"> </span>this_pat
  clone_tbl_all_pats_all_trees_list[[cnt]] &lt;-<span class="st"> </span>clone_tbl_this_pat_all_trees 
}

<span class="co"># Now we aggregate the list of tibbles into one big tibble</span>
clone_tbl_all_pats_all_trees &lt;-<span class="st"> </span><span class="kw">do.call</span>(<span class="st">&quot;rbind&quot;</span>, clone_tbl_all_pats_all_trees_list)</code></pre></div>
<p>The herewith generated <code>avg_rates_m</code>, <code>list_of_num_trees_all_pats</code>, <code>list_of_clon_excl_all_pats</code>, and <code>clone_tbl_all_pats_all_trees</code> will be used as input to the functions used in steps 2 and 3 of this vignette.</p>
<hr />
</div>
<div id="generate-null-distribution-of-test-statistic" class="section level2">
<h2>2. Generate null distribution of test statistic</h2>
<p>In order to generate the null distribution of the test statistic, the empirical cumulative distribution function (ECDF) of the test statistic is computed. The null hypothesis assumes independence of mutations, and therefore states that the gene pairs occur across the cohort in the clonal exclusivity pattern as often as expected by chance according to the overall rates of clonal exclusivity. The function to compute the ECDF of the test statistic under the null <code>generate_ecdf_test_stat</code> takes as input from all patients the data computed in step 1, namely the average rate of clonal exclusivity, as well as the histogram of how often pairs occur in the tree, and how often they were clonally exclusive. Here, it is assumed that step 1 of this vignette was run for all patients, and that you have the parameters <code>avg_rates_m</code>, <code>list_of_num_trees_all_pats</code>, <code>list_of_clon_excl_all_pats</code>, and <code>clone_tbl_all_pats_all_trees</code>. If not, you can load the precomputed data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># load the data as generated in step 1 from all 82 patients</span>
<span class="kw">data</span>(<span class="st">&quot;clone_tbl_all_pats_all_trees&quot;</span>)
<span class="kw">data</span>(<span class="st">&quot;avg_rates_m&quot;</span>)
<span class="kw">data</span>(<span class="st">&quot;list_of_num_trees_all_pats&quot;</span>)
<span class="kw">data</span>(<span class="st">&quot;list_of_clon_excl_all_pats&quot;</span>)</code></pre></div>
<p>Now we can have a look at this data.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># the tibble containing the gene-to-clone assignments from all patients and the whole collection of trees</span>
<span class="kw">print</span>(clone_tbl_all_pats_all_trees)
<span class="co">#&gt; # A tibble: 14,108 x 11</span>
<span class="co">#&gt;    file_name  patient_id altered_entity clone1 clone2 clone3 clone4 clone5</span>
<span class="co">#&gt;  * &lt;chr&gt;      &lt;chr&gt;      &lt;chr&gt;           &lt;int&gt;  &lt;int&gt;  &lt;int&gt;  &lt;int&gt;  &lt;int&gt;</span>
<span class="co">#&gt;  1 cloe_seed~ 01         ENSG000000081~      0      1      1      1      0</span>
<span class="co">#&gt;  2 cloe_seed~ 01         ENSG000000511~      0      1      1      1      0</span>
<span class="co">#&gt;  3 cloe_seed~ 01         ENSG000000784~      0      0      1      1      0</span>
<span class="co">#&gt;  4 cloe_seed~ 01         ENSG000000847~      0      1      1      1      0</span>
<span class="co">#&gt;  5 cloe_seed~ 01         ENSG000001001~      0      1      1      1      0</span>
<span class="co">#&gt;  6 cloe_seed~ 01         ENSG000001059~      0      0      1      1      0</span>
<span class="co">#&gt;  7 cloe_seed~ 01         ENSG000001082~      0      0      1      1      0</span>
<span class="co">#&gt;  8 cloe_seed~ 01         ENSG000001088~      0      1      1      1      0</span>
<span class="co">#&gt;  9 cloe_seed~ 01         ENSG000001112~      0      0      1      1      0</span>
<span class="co">#&gt; 10 cloe_seed~ 01         ENSG000001137~      0      1      1      1      0</span>
<span class="co">#&gt; # ... with 14,098 more rows, and 3 more variables: clone6 &lt;int&gt;,</span>
<span class="co">#&gt; #   clone7 &lt;int&gt;, tree_id &lt;int&gt;</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># the average rates of clonal exclusivity from each patient</span>
<span class="kw">head</span>(avg_rates_m)
<span class="co">#&gt;          01          02          04          05          06          07 </span>
<span class="co">#&gt; 0.005911330 0.184444444 0.000000000 0.300000000 0.037706767 0.004545455</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># the histograms of how often pairs were mutated across the collection of trees - here printed for patient 02</span>
list_of_num_trees_all_pats[[<span class="dv">2</span>]]
<span class="co">#&gt;  [1] 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20</span>
<span class="co">#&gt; [24] 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># the histograms of how often pairs were clonally exclusive across the collection of trees - here printed for patient 02</span>
list_of_clon_excl_all_pats[[<span class="dv">2</span>]]
<span class="co">#&gt;  [1]  5  0  0  5  5  5  4  4  6  0  0  0  0 18  1  1  1  0  1  0  0  2  2</span>
<span class="co">#&gt; [24]  2  0  0  0  0  0  0  0 19  1  1  0 17  1  1  1 20 20 19  0  2  2</span></code></pre></div>
<p>For each number of patients, a pair can be mutated in, an ECDF is built separately. With the parameter <code>num_pat_pair_max</code>, it can be decided up to which number to compute it. You can use the function <code>pairs_in_patients_hist</code> to find out what is the maximum number of patients that pairs are mutated in. All unique gene pairs are efficiently computed using functionalities from the R package <code>caTools</code> [<a href="https://CRAN.R-project.org/package=caTools">6</a>].</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">pairs_in_patients_hist</span>(clone_tbl_all_pats_all_trees)
<span class="co">#&gt; Found 82 different patient id's in the provided tibble.</span>
<span class="co">#&gt; There are in total 6270 pairs of genes/pathways mutated in at least one patient.</span>
<span class="co">#&gt; There are in total 325 pairs of genes/pathways mutated in at least 2 patients.</span>
<span class="co">#&gt; The maximum number of patients that a pair is mutated in is 15. Please run the function 'generate_ecdf_test_stat' with num_pat_pair_max = 15.</span>
<span class="co">#&gt; # A tibble: 8 x 2</span>
<span class="co">#&gt;   pairs_count patient_count</span>
<span class="co">#&gt;         &lt;int&gt;         &lt;int&gt;</span>
<span class="co">#&gt; 1        5945             1</span>
<span class="co">#&gt; 2         274             2</span>
<span class="co">#&gt; 3          35             3</span>
<span class="co">#&gt; 4           9             4</span>
<span class="co">#&gt; 5           4             5</span>
<span class="co">#&gt; 6           1             7</span>
<span class="co">#&gt; 7           1             9</span>
<span class="co">#&gt; 8           1            15</span></code></pre></div>
<p>In this data set, a total of 5’945 pairs are mutated in just one patient, and 274 pairs are mutated in two patients. GeneAccord will only consider pairs that occur in more than one patient. One pair is mutated in 15 patients, which is the maximum number. So, in practice, <code>num_pat_pair_max = 15</code> should be chosen. For the vignette, to demonstrate the functionality of the package, we set it to three such that it runs faster.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">num_pat_pair_max &lt;-<span class="st"> </span><span class="dv">3</span></code></pre></div>
<p>The following command will generate the ECDF of the test statistic under the null for this data set. The number of pairs that are simulated is set to 1,000. This step may take quite some time.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">1234</span>)

ecdf_list &lt;-<span class="st"> </span><span class="kw">generate_ecdf_test_stat</span>(avg_rates_m, list_of_num_trees_all_pats, list_of_clon_excl_all_pats, num_pat_pair_max, <span class="dt">num_pairs_sim =</span> <span class="dv">1000</span>)
<span class="co">#&gt; There were 82 rates of average clonal exclusivity provided.</span>
<span class="co">#&gt; The average rates will be distorted with a beta distribution and M = alpha + beta = 1000.</span>
<span class="co">#&gt; Additionally, from the 82 patients, the values from the data for all pairs of # clon. excl./ # occurrence across all trees are provided.</span>
<span class="co">#&gt; From those, it will be sampled 1000 times to obtain the test statistic under the null.</span>
<span class="co">#&gt; This procedure is done for num_patients = 2, ... 3.</span>
<span class="co">#&gt; Generate ECDF for the case where a gene pair is mutated together in 2 patient(s).</span>
<span class="co">#&gt; Generate ECDF for the case where a gene pair is mutated together in 3 patient(s).</span></code></pre></div>
<p>In practice, it is recommended that the parameter <code>num_pairs_sim</code> is set to a higher value, to get most accurate results. Here, as a demonstration, it was set to 1,000 such that it runs faster. However, we advice to choose rather a higher number in your analysis, e.g. 100,000. The function then takes much more time, but it could be run over night on a computing cluster. For the purpose of this vignette, the pre-computed ECDF can be loaded, which was run with <code>num_pairs_sim = 100000</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(<span class="st">&quot;ecdf_list&quot;</span>)</code></pre></div>
<p>The distribution of the test statistic under the null will later be used in the clonal exclusivity test to evaluate how unlikely an actually observed test statistic is.</p>
<hr />
</div>
<div id="clonal-exclusivity-test" class="section level2">
<h2>3. Clonal exclusivity test</h2>
<p>Now, we’re ready to perform the clonal exclusivity test to identify pairs of genes that are mutated in different clones more often than expected. To this end, we call the function <code>GeneAccord</code>. The required input parameters are the tibble containing the gene-to-clone assignments from all patients and all trees, the average rates of clonal exclusivity, and the ECDF’s as computed in step 2 of this vignette. This function makes use of functionalities from the R packages <code>gTools</code> [<a href="https://CRAN.R-project.org/package=gtools">7</a>], and <code>maxLik</code> [<a href="https://CRAN.R-project.org/package=maxLik">8</a>].</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># perform the clonal exclusivity test</span>
res_pairs &lt;-<span class="st"> </span><span class="kw">GeneAccord</span>(clone_tbl_all_pats_all_trees, avg_rates_m, ecdf_list)
<span class="co">#&gt; There were 20 different tree inferences per patient.</span>
<span class="co">#&gt; 14108 alterations in total found in the provided tibble.</span>
<span class="co">#&gt; Found in total 362 different mutated genes/pathways in the provided tibble.</span>
<span class="co">#&gt; Found 82 different patient id's in the provided tibble.</span>
<span class="co">#&gt; A total of 325 pairs of genes/pathways is mutated in at least one patient.</span>
<span class="co">#&gt; And 325 pairs of genes/pathways are mutated in at least 2 patients.</span>
<span class="co">#&gt; The maximum number of patients that a pair is mutated in is 15, which is higher than the number of ECDF's provided. This may be problematic.</span>
<span class="co">#&gt; If you have more time, please re-run the function 'generate_ecdf_test_stat' with num_pat_pair_max = 15.</span>
<span class="co">#&gt; Otherwise, you can also wait and see if one of the pairs that will be tested is actually mutated in so many patients.</span>
<span class="co">#&gt; Maybe all pairs tested will be mutated in less patients, and you're fine.</span>
<span class="co">#&gt; 325 gene/pathway pairs will be processed.</span>
<span class="co">#&gt; 325 pairs were mutated in more than one patient and for them the delta was computed.</span>
<span class="co">#&gt; There were 16 pairs where delta is &gt; 0 and for which the p-value was computed.</span></code></pre></div>
<hr />
</div>
<div id="examining-and-visualizing-results" class="section level2">
<h2>4. Examining and visualizing results</h2>
<p>The results tibble now contains all pairs of genes that are mutated together in at least two patients, and whose delta parameter was greater zero, indicating that they tend to be clonally exclusive. See [1] for more details about the parameter delta. We can extract the significant pairs using the pipe command <code>%&gt;%</code> from the R package <code>magrittr</code> [<a href="https://CRAN.R-project.org/package=magrittr">9</a>], and using the function <code>filter</code> from the R package <code>dplyr</code> [<a href="https://CRAN.R-project.org/package=dplyr">12</a>].</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">suppressMessages</span>(<span class="kw">library</span>(magrittr))
<span class="kw">suppressMessages</span>(<span class="kw">library</span>(dplyr))

<span class="co"># extract all pairs with an adjusted p-value less than 0.05</span>
sig_pairs &lt;-<span class="st"> </span>res_pairs <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(qval <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>)

<span class="co"># print the top pairs</span>
sig_pairs[<span class="kw">order</span>(sig_pairs<span class="op">$</span>pval),]
<span class="co">#&gt; # A tibble: 5 x 7</span>
<span class="co">#&gt;   entity_A  entity_B  num_patients pval    mle_delta test_statistic   qval</span>
<span class="co">#&gt;   &lt;fct&gt;     &lt;fct&gt;            &lt;int&gt; &lt;fct&gt;       &lt;dbl&gt;          &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt; 1 ENSG0000~ ENSG0000~            2 0.0009~     161.            5.73 0.0147</span>
<span class="co">#&gt; 2 ENSG0000~ ENSG0000~            2 0.0098~      39.2           3.88 0.0434</span>
<span class="co">#&gt; 3 ENSG0000~ ENSG0000~            2 0.01085      52.5           3.78 0.0434</span>
<span class="co">#&gt; 4 ENSG0000~ ENSG0000~            2 0.01085      52.5           3.78 0.0434</span>
<span class="co">#&gt; 5 ENSG0000~ ENSG0000~            2 0.01549      24.6           3.43 0.0496</span></code></pre></div>
<p>We notice that the gene identifiers are ensembl gene id’s. In order to find out the corresponding gene names, you can use the function <code>map_pairs_to_hgnc_symbols</code>. First, a tibble containing gene id conversions needs to be created. The function to do so, <code>create_ensembl_gene_tbl_hg</code>, makes use of functionalities of the Bioconductor R package <code>BiomaRt</code> [<a href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html">10, 11</a>].</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># this will create a tibble with ensembl gene id's and corresponding hgnc gene symbols of the human genes</span>
all_genes_tbl &lt;-<span class="st"> </span><span class="kw">create_ensembl_gene_tbl_hg</span>()</code></pre></div>
<p>You can also load it from the pre-computed data of this package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(<span class="st">&quot;all_genes_tbl&quot;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># get the hgnc symbols of the significant pairs</span>
sig_pairs &lt;-<span class="st"> </span><span class="kw">map_pairs_to_hgnc_symbols</span>(sig_pairs, all_genes_tbl)

<span class="co"># print the top pairs - here just showing the hgnc gene symbols and not the ensembl gene id's </span>
sig_pairs[<span class="kw">order</span>(sig_pairs<span class="op">$</span>pval),] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(<span class="op">-</span>entity_A, <span class="op">-</span>entity_B, <span class="op">-</span>num_patients)
<span class="co">#&gt; # A tibble: 5 x 6</span>
<span class="co">#&gt;   pval             mle_delta test_statistic   qval hgnc_gene_A hgnc_gene_B</span>
<span class="co">#&gt;   &lt;fct&gt;                &lt;dbl&gt;          &lt;dbl&gt;  &lt;dbl&gt; &lt;fct&gt;       &lt;fct&gt;      </span>
<span class="co">#&gt; 1 0.0009200000000~     161.            5.73 0.0147 TP53        MUC16      </span>
<span class="co">#&gt; 2 0.0098300000000~      39.2           3.88 0.0434 TP53        BAP1       </span>
<span class="co">#&gt; 3 0.01085               52.5           3.78 0.0434 CACNA1S     MAP3K3     </span>
<span class="co">#&gt; 4 0.01085               52.5           3.78 0.0434 CSMD1       MAP3K3     </span>
<span class="co">#&gt; 5 0.01549               24.6           3.43 0.0496 USP36       CNOT1</span></code></pre></div>
<p>The HGNC gene symbols are now displayed in two additional columns. We are also interested in which patients these pairs are mutated and clonally exclusive.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># find out in which patients they occur and are clonally exclusive</span>
sig_pairs &lt;-<span class="st"> </span><span class="kw">take_pairs_and_get_patients</span>(clone_tbl_all_pats_all_trees, sig_pairs)
<span class="co">#&gt; Found 5 pairs of interest for which the patients will be extracted.</span>
<span class="co">#&gt; There are 82 different patients in the clone tibble and the number of tree inferences is 20.</span>
<span class="co">#&gt; A pair is counted as clonally exclusive in a patient, if it clonally exclusive in the majority of tree inferences, i.e. in at least 11 trees of a patient.</span>

<span class="co"># print the top pairs</span>
sig_pairs[<span class="kw">order</span>(sig_pairs<span class="op">$</span>pval),] <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(<span class="op">-</span>entity_A, <span class="op">-</span>entity_B, <span class="op">-</span>mle_delta, <span class="op">-</span>test_statistic, <span class="op">-</span>num_patients)
<span class="co">#&gt; # A tibble: 5 x 6</span>
<span class="co">#&gt;   pval          qval hgnc_gene_A hgnc_gene_B mutated_in clonally_exclusiv~</span>
<span class="co">#&gt;   &lt;fct&gt;        &lt;dbl&gt; &lt;fct&gt;       &lt;fct&gt;       &lt;chr&gt;      &lt;chr&gt;             </span>
<span class="co">#&gt; 1 0.00092000~ 0.0147 TP53        MUC16       05;81      05;81             </span>
<span class="co">#&gt; 2 0.00983000~ 0.0434 TP53        BAP1        30;84      84                </span>
<span class="co">#&gt; 3 0.01085     0.0434 CACNA1S     MAP3K3      16;36      36                </span>
<span class="co">#&gt; 4 0.01085     0.0434 CSMD1       MAP3K3      16;36      36                </span>
<span class="co">#&gt; 5 0.01549     0.0496 USP36       CNOT1       16;49      49</span></code></pre></div>
<p>The top pair is mutated in patients 05 and 81, and is clonally exclusive in both of them. The results tibble can be saved as a tab-separated file.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># define output file</span>
sig_pairs_tsv &lt;-<span class="st"> &quot;GeneAccord_sig_pairs.tsv&quot;</span>

<span class="co"># save the results into the tsv-file</span>
sig_pairs_table &lt;-<span class="st"> </span><span class="kw">write_res_pairs_to_disk</span>(sig_pairs, avg_rates_m, sig_pairs_tsv)</code></pre></div>
<table>
<thead>
<tr class="header">
<th align="left">Gene A</th>
<th align="left">Gene B</th>
<th align="right">P-value</th>
<th align="right">Adjusted p-value</th>
<th align="left">Mutated in (rate)</th>
<th align="left">Clonally exclusive in</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">TP53</td>
<td align="left">MUC16</td>
<td align="right">0.00</td>
<td align="right">0.01</td>
<td align="left">Patient 05 (0.3); Patient 81 (0.19)</td>
<td align="left">Patient 05; Patient 81</td>
</tr>
<tr class="even">
<td align="left">TP53</td>
<td align="left">BAP1</td>
<td align="right">0.01</td>
<td align="right">0.04</td>
<td align="left">Patient 30 (0); Patient 84 (0.14)</td>
<td align="left">Patient 84</td>
</tr>
<tr class="odd">
<td align="left">CACNA1S</td>
<td align="left">MAP3K3</td>
<td align="right">0.01</td>
<td align="right">0.04</td>
<td align="left">Patient 16 (0); Patient 36 (0.15)</td>
<td align="left">Patient 36</td>
</tr>
<tr class="even">
<td align="left">CSMD1</td>
<td align="left">MAP3K3</td>
<td align="right">0.01</td>
<td align="right">0.04</td>
<td align="left">Patient 16 (0); Patient 36 (0.15)</td>
<td align="left">Patient 36</td>
</tr>
<tr class="odd">
<td align="left">USP36</td>
<td align="left">CNOT1</td>
<td align="right">0.02</td>
<td align="right">0.05</td>
<td align="left">Patient 16 (0); Patient 49 (0.18)</td>
<td align="left">Patient 49</td>
</tr>
</tbody>
</table>
<p>The ECDF’s of the test statistic under the null hypothesis can also be visualized.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># plot the ECDF of the test statistic under the null</span>
<span class="kw">plot_ecdf_test_stat</span>(ecdf_list)
<span class="co">#&gt; Found 2 ECDF's to plot.</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAABIAAAAHgCAMAAADquuaVAAAA21BMVEUAAAAAADoAAGYAOjoAOmYAOpAAZmYAZpAAZrY6AAA6ADo6AGY6OgA6Ojo6OmY6ZpA6ZrY6kLY6kNtmAABmADpmAGZmOgBmOjpmOpBmZgBmZmZmkLZmkNtmtttmtv+QOgCQOjqQOmaQZgCQZjqQkGaQkLaQtpCQttuQtv+Q27aQ29uQ2/+zs7O2ZgC2Zjq2ZpC2kDq2kGa225C227a229u22/+2/7a2/9u2///T09PbkDrbkGbbtmbbtpDb27bb29vb2//b/9vb////tmb/25D/27b//7b//9v///8aYIoyAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAgAElEQVR4nO2dC3vrxnVFIfmqpVI3ccXW6ctiG7d122umTRMnIhOnqXgvyf//i4oHHyA1oIg5czDngGt9nyxeibMxGGxtzBkAdLEFAMhEkbsDAHC7EEAAkA0CCACyQQABQDYIIADIBgEEANkggAAgGwQQAGSDAAKAbBBAAJANAggAskEAAUA2CCAAyAYBBADZIIAAIBsEEABkgwACgGwQQACQDQIIALJBAAFANgggAMgGAQQA2SCAACAbBBAAZIMAAoBsEEAAkA0CCACyQQABQDYIIADIBgEEANkggAAgGwQQAGQjdQAtiz0Pr+U/N3/4snz50x9av7r72Q8nb7x/2TVdT+8+budF8Vy/LibvbGl+bHkVm18+FsVXl9/yXy/bzazpefc7InkrfHlTV3ago0eH0bl2I59/vjtQnx77jWta8E9nU+P++bY8NP/w2ts/ugG0/nr3j6f2r+6etyEDLasm813D9AZa7fvRyeev798xUP2OSJIY6E0HunrU10Crx+bQfKwaXBwlXfBPF7b9Uw54xWTb1z+qAVT2Zc9z+1elzd8aqOn4vHmvgoGW9WZlin23mZw3HejqUc+eloN//7E8eVVHbdn/vJoO/KOKln+qw/FN9Z/nbU//pA+gVsfLDn34rkzZWdEYu/7VH7+ug/LN8VzVP5g3lmoM1IxC+fqp/sGPXxZ332x//1h8qFqWv/x1WTY0E/LNL0tbfvVab/Pu374u7n/Yqf5YvuWLamY43zv32PRD/aYfq7PsT3ebLu5/1yT+ieB/lv/4cHjHy/b3Xx4qge2JxPaCfHMqaXdvd3I5VQvu6H4U9h046/NLq7t1qfDhu7MzWGsnaj49tv6Wq63+/HGy31T5u+fYwy8G/7j0z+ZXf1m+q3FOP/9oBlB1Wq3/sfmnr/50/FU5FM0ufQy0nDcT3ZCBGr7cn/XmrTPg7rcfXvZnxn0GLw9n01MDHUdvdXjZNtAbwdY7Xpanp96WxPaC/P5IHrvX/slBLbSj5wY67/NLq7uHWcOZgU77eG6gQ8+bP958NRj+8eufzbz+Zz//6JVgz+fz4IO36hfL4/tqyh2t3lyO1t81IzJ5Y6Cn7R+K3UyvmWw//FD9pHnjD9WgPNUdKH/8v43ovqZoftE+wve7puVm//x117TeXDPgp4LluWi/zd3Z9bVq/XTo+UHigvzhSB66V//kXC20o4dROHTxtM/t7pbmmrxWJ7RzAx12YjcwgQBqhqtUy1eD4R+3/lntZ0i9/KMZQK0xbX51MFDZ7XMD7dxWDsX/lM1CBmpOLA/HobvbTzB3mbs35nECuJtIzptftA30fPzt9o+/+rI4bG5/VM8Ey23u31G+/PCbs93eS1yQPxzJQ/fqn5yrhXb0zEBv+tzu7uHd5wY67EQnlXfbHc8B/nHrn+qIfNUateswE0C7N1edL63xH6Eafu+y9rjWYu08bk/N9yv4y7NftJpuN982TSdtA70VbDm6maTWCwO7zRwlLsgfiunT7p2rde1o20DnfW51dzcROJqgvdnLS7Ofp0VruDKBf9z6Z7v98XEXZD38o7kG1D2FPjvQFS0DlS//rGWg+jehcT2YY1WcH+/TDqwqm741UPXj6vh9+Nc/Tk8N9FawfUr93Fwcvvv3Rq0tcUG+w0Dnal0GOp3ln/a51d39Lr9joMAUuvpRM3XOG0D4x6d/jq3MBFBrEbFapN//qvnxBQM1Z7fJuwZqn8Gejx3odQZrtNZnBnor2DZQaaF/qZb4dpVuW+KCfJeBztSuMNCbPre6u9fsbaBj/pgJIPzjyD+tnpsJoG3wMmr1r8nbXrYNVC/KH96zKjoMdKiU26fKE933a/jVvrg9WUR8K3hqoJLNL/av2xIX5LsNdKL2dkdbo3A4YZ32ud3drhr+HQOtp+1LPzYCCP+48c/nnz82o191KHMA7TlUuzWBG8lOe3lcRNxdupzU355qw3UY6P54FePum2oN4/zy7PtXMep3tD29v4pxJthsuX5dHr+/qme+xzPYQeKCfIeBztVCBjqMQtP8TZ/b3W2uYszfXMV4p4afH/PHyCI0/nHkn/Jd5Rt21+KMLELXQbq/lf6r9q92t9KfGuh4GXV3X8aklbMdBjpuZ9phzON9HGcGeuPxSXPPxel9HEfBZsv1O15+eTp5aEtckO86g52pvd3R1igcunja55dWd0/GZXutgQ7buH85rI5kAf+49M9+I7vyzcRl+HoHdg8T/qb1q8PDhGfztOXuHpT9Atyk/PapNODPft15FeN4J+vnbx93L8906ztZ//Fsg2XT3327e391ReDDd/X0vrr74cNvmnntuWCz5fod5cnp+ITk9lTignznFPpU7e2OtkZh14HzPv/Q6u5289+PxRff9LyK0X6wYZPzYTD849I/zcPM9cOoPf1j5+M4VsOtPCjXGDlLmAR8esy3BCQA/xihn3/sBFAZnEM9goSBLrHKWIEJwD9G6OcfOwFUTvMmA20JA11inrECk4B/bNDPP4YCaD0datwx0AXyfiCZAPxjgswfSAYAcDUEEABkgwACgGwQQACQDQIIALJBAAFANgggAMgGAQQA2SCAACAbBBAAZIMAAoBsEEAAkA0CCACyQQABQDYIIADIBgEEANkggAAgGwQQAGSDAAKAbBBAAJANAggAskEAAUA2CCAAyAYBBADZIIAAIBsEEABkgwACgGwkDqACRkJaX+CfW+PqI97LH5tZo975v59nQjUSdA4k/rkVdAJoWTw1L1b7FyI5sIvKgcQ/N4NKAG1mB9ssH17Fctvtote77UqY6ETa/dBIAvwzdomjgkoArafP+5erjkn0bRrIRCfsBxD+GbuEcgAlP4OBWVzMgMAsWmtAu1MYNfzYUVoDwj83gtJVsPW0uYrRcf661Sm0iU7YL8Hwz9gllEuw5HImBo0ACinkug+o17tNDBoSIQUfAQRmcRFAYBbtAFomuYoBZlE+kPhn5Aw8AzrcgL1YfP/9opyIXfd9ceX7rH8X78du3IrM3xetfyfxBf657vtCrrMI/HxwHy36+4c1oOskLj73spA/OWNiKFgDkkgIDr/cQEGJwceCNaA0hA5w7j4NjIsAys5NW+QiBND7XDid5O5afgigS2CU9/ARQFmm0G9ihsvwAQUXATT4oAWDx8ShsyGhXILtbyOr8PcsT+dZiwAKKOg8C+bTP+9MjU0cOhsS2mtAm1nnB7nEyA0CdVUUKqPlzD84Jx61DySbpJTTg/UcIVofSDbJsdneYB0pWmtAq+L54u9zT6FjYocSLKCg9Mdn3T97B+XuhVMJb5fh0w2aZL5DAAUUbnIROs5BJg6dDQlvAZRgexRaOrgIoKRbxkUpGXkAscKjzU0FEE5Kjo8A6jvnC+SOsZmnb4lbLMHk4WPi0NmQ8FaCXbfHF69oGRt33xI3FkCVoxbyHTVx6GxIeAugK/SYHOfBRQBJNoSxVBlDAOGRjIw5gDCWPj4C6MKc71qPGJt5+pa4hRLszFgmxn00Et5KsM49vv4MZWzcfUuMPYACUx8T4z4aCW8B1CXCBDk7LgKojzB116B4DiB8YoBRBRDZMzg+Aig453P4II6JTlCCdWlecpSJcR+NhLcSLLTHPc9Vxsbdt8QYA+idyY+JcR+NhLcACgkwVzaBiwC6Qg8/5cFrAOEXI4wjgLBTLnwE0Ns5X+/uGJt5+pYYWQl2zfTHxLiPRsJbCRb4NKi+GzQ27r4lRhVA11VfJsZ9NBLeAih1c0iGiwC6qISXcuIzgDCNGbwHEFbKi48AOp/zRXTG2MzTt8R4SjBfj/KMRsJbCUYA2ZIYTQBdP/8xMe6jkfAWQOeNmTabwUUAactAND4DKFUvQIzrAOJMlh0fAXQ254vpi7GZp2+JkZRgffLHxLiPRsJbCXa6x5n+p14EUEjBdQDJJRL04hYlvAVQuraQGBcB1KGBkfJDAIEIzwGUQAOE+AigsxIsZoPGZp6+JUZRgjn8PKnRSHgrwU72OG7mbGzcfUuMI4DkEv1AIqDgI4BSNYXkuAigoAI+sgABBCL8BlCCXoAYHwF0WoJFbdDYzNO3xAhKsL4TIBPjPhoJbyUYAWRLYgwB1HODJsZ9NBLeAuikJXNnS7gIIA0BSILDAErYCxDjNIA4jRnBRwC153yRHTE28/Qt4b8E691tE+M+GglvJRgBZEuCAIoBiYCCjwBK0xIUcBFA6dtDIpQCaF4Uk82sKIrnJHLthljHFDqHQ88/u+a4yAg6AbR8eC09NNlu19OnBHIppvzGZp6+JbRLMEX/xDQPSvQHiYCCSgBtZqVtVncft42VpHIEkDUJ5QDS9E9M86BEf5AIKKgE0HpazpxX9y/b/X9lcmkaggoax0PTP0maQzJ8zIBSNAQVXMyA3rTGRVZQXgOqrSSWa8/YKMEMSAy2BpTcPzGtgxIRIBFQ8HEVLKK/3RLREEABBRdXwQggYxJ+7wNi7mwMl/cB4SIzEEAgggACCdoBtExyFeMwY4tfPTQ28/QtMdyjGIn9E9U6KBEBEgGFgWdAxZ5F2YXrvxb71z3bWftaZN6+zn4MiNQ/u6/CwBhGfS1SHzsL+3HloU/spEGbgRoeSzBcZAdna0BYxxoEEEhQCqD6EmpJRwUfW8MLbiBLUCrIJUx0wsMakJJ/ohoHJWJAIqCgdCNisbt/bFUkvZFM4HZj4+5bQjuAtPwT1TgoEQMSAQW9RzEa0t5Kz9zZHGqPYjSoPIqBjeyg9zBqQ9qHCXGOOdQeRm3QeBiVJ8EM4WMGRAlmS0K5BFPzT0zboEQUSAQUtNaAdqcw1oDsdcJ+AKn5J6ZtUCIKJAIKSlfB1tPmKkbH+YsSbDToHBIl/8jbQmJc3QdE8W4Ph/cB4SJD+AighbwPxmaeviXU7wN6F0ow1xLqJVhiOQLIlgQBhEQiBR8BpNMHSICLADptio0MQQCBCIcBlK4XIMZHAFGC2ZJwXYLF9djEuI9GwlsJRgDZkiCAkEik4COAdPoACXARQKmaQnIIIBBBAIEEHwFECWZLghIMiUQKBNCQEiY6ceMBFHkV3sS4j0bCWwDp9AES4CKA0rQEBQggEEEAgQQfAUQJZkvCdQkWt0ET4z4aCW8lWNNf0T30xsbdtwQBhEQiBR8BpNIFSIGLAErTEhQggEAEAQQSfATQQt4FYzNP3xKUYEgkUiCAhpQw0YnbDqDYhUQT4z4aCW8BpNIFSIGLAErREFQggEAEAQQSfAQQJZgtCc8lWOQGTYz7aCS8lWAEkC0JAgiJRAo+AkilC5ACFwGUoiGoQACBCAIIJPgIIEowWxKUYEgkUiCAhpQw0YmbDqDo5wlNjPtoJLwFkEoXIAUuAkjeDpQggEAEAQQSfAQQJZgtCcclWOwGTYz7aCS8lWAEkC0JAgiJRAo+AqhpwvTZIC4CSN4OlPAUQGl7AEkggECCjwBayHtgbObpW8JtCRY/izYx7qOR8FaCEUC2JPwGUPQGTYz7aCS8BZBGDyAJLgJI2gzUUAqg+f3L9tNjUdx9TCIX2wTU0TkqGv4RNAM1dAKo9s9PSvOsp88J5CjBjElol2Aq/unfLCgRDxIBBZUAWk+fShNNqpfLh1exHAFkTUI5gHT8079ZUCIeJAIKSgH0vN3MnqqXq/JcJpWLbwLq6ASQhn8EzUANpRJsUp67JtWrNGew2Cagjk4JNlHwj6AZqKETQOvp/Ut9Clt1rSJSgnmW0F4DUvHPVnQzvYlxH42E/mX4VVEzSSNHANmS0L8PSME/vVsFJQQgEVDgPiAQ4eo+ICxkDgIIRBBAIEE7gJZJrmJQgtmSGO5RjIT+6d0qKCEAiYDCwDOgYs+i7ML1X4vqe882Fr8WBvqQfj8GROKf8qswMHair0XqY2dhP6489Imd1L8F82eLUIKBBD9rQJjHJAQQSFAKoM2smSl3VPAxNbzQPAlKBbmEiU54WANS8E//VkEJAUgEFHQCaFk8NS9W+xciOQLImIR2AGn4p3+roIQAJAIKKgG0e46nIt2t9EyfTaJxWFT8E98KFFF7GHVHuocJMY9JtB5G3ZH2YVQ8ZA4fMyBKMFsSyiWYin/6twpKCEAioKC1BrQ7hbEGZK8T9gNIxT/9WwUlBCARUFC6CraeNlcxOs5flGCjQeewKPgnvhUown1AIMLTfUDcy2qPiADafVJC0fFpvYm3W0MJZktCVIJl8k/vRkEJCUgEFPoGUDk3njSvNrPu28SulrsSAsiWRHwA5fNP70ZBCQlIBBR6BtD6b9qeOf1XLyjBRkLPtcF8/oltBKq4WQOifreJqzWg1L0AMf0DaPOnQbdbs5Cbx9jM07eEZA0ok3/6NwpKSEAioNA/gNbTSf1988/xFTwB5FtCEkCZ/NO/UVBCAhIBhairYNX/qmApWUKMKcEEGwM9Yq6C5fBPdCNQJWYNqPmshI5bVFNvN7oBDELEccnin+hGoErUInTlINFdHJRgviWEj2Lk8E//RkEJCUgEFGICaF4U38y675LvK3cFBJAtCVkAZfFP/0ZBCQlIBBRiFqHr2fOKNSDYRi1CZ/FPdCNQJSKA/rY5dQ15FSOmAQxC/wDK45/oRqCKjxsRKcFsSWh/HMf7UIK5lvD2KAYBZEsiOoDy+ad/o6CEBCQCCj4eRo1pAIPg4mFUSSNQxcfHccQ0gEFw8XEckkagCmtAQ0qY6ARrQDGYGPfRSKh/JGtiOQLIlgQBhEQihYgAevf/WtlPTqsBDEL/45LHP9GNQJWIAJqLnNN3u/v3Yx6b9D8uWfxTt8FD9oi5E1r2GGHP7dYs5CcvYzNP3xKyj+PI4Z/+bYISIpAIKMQEkPA5wn7brSGAbEnIAiiHf/q3CUqIQCKgEPGJiLOJePsRJViCbYIC/T8RMYt/YtuAMlH3ASU4hRFAIyHmPqAM/oltA8pEPQ0/+FUMSjBbErISLId/+rcJSohAIqDg5T4gAsiShM/7gAggMxLeAojps1lcBFB8G1Cm79PwSS5h9Nlu5PthIHo+DZ/LP7FtQJmYANrM5HdyUIJ5loj/OI5c/unfJighAomAAgE0pISJThBAMZgY99FIeAsgps9mcRFA8W1AGQIIRBBAIMFHAFGC2ZKgBEMikQIBNKSEiU4QQDGYGPfRSAgCqDgw6Oe5MH02St8AyuSf2DagjOqNiJ8eO+/6IIBGguaBSeif2DagjEoAXXGeowTzLKH8KIaCf/q3CUqIQCKgoDMDWtX/9910ZzACyJaE9rNg6f3Tv01QQgQSAQWlEmw9fXilBLsFdA5Mcv/EtgFl1NaA5ncfCaAbQOvAJPZPbBtQRm8Relk8UYJZ7ISDEqwirX+qJpRgZiQG+TiOT49fEEAGO+EkgNL6p3eToIQMJAIKmpfhN7PuD9+kBBsJigcmpX/imoA6fCAZiPDzgWRYyCLaAbTkPiBrnfBSgtWk8k/vJkEJGUgEFAaeAR3uL1uUXbj+a9HjvZa/xrkfAyLxT882Jr8WqY+dhf248tAndpLy+2EgKMFAAmtAIIIAAglKAbSZvfPEM2tAniVaCjp/1sn907tJUEIGEgEFnQBaFrtPfFkVHR/9QgB5ltAOoPT+6d0kKCEDiYCCSgC1PnFqWT3TI5SLeT8MhMaBUfBPXBNQR+njOA43kK2SXEbFPWbR+TiO5P6JawLq+JgBUYLZklAuwRT807tJUEIGEgEFrTWg3SmMNSB7nbAfQAr+6d0kKCEDiYCC2ucBNVcxOs5flGCjQevzgBL7J64JqMN9QCDCzX1Ask/jACV8BBAlmC0J9fuA3qV/CSbsqIlxH42EegmWWI4AsiVBACGRSMFHAFGCmcVFAMW1gAEggEAEAQQSfAQQJZgtCUowJBIpEEBDSpjoBAEUg4lxH42EtwBiAm0WFwEU1wIGgAACEQQQSPARQJRgtiQowZBIpEAADSlhohMEUAwmxn00Et4CiAm0WVwEUFwLGAACCEQQQCDBRwBRgtmSoARDIpGCkwCSP8psbNx9SxBASCRS8BFAzJ/N4iKA4lrAABBAIIIAAgk+Amgh37qxmadvCUowJBIpEEBDSpjoxG0GkHQV0cS4j0bCWwAxfzaLiwCKaQCDQACBCAIIJPgIIEowWxIeSzDhBk2M+2gkvJVgBJAtCQIIiUQKPgKICbRZXARQTAMYBAIIRBBAIMFHAFGC2ZKgBEMikQIBNKSEiU4QQDGYGPfRSHgLICbQZnERQDENYBAIIBBBAIEEHwFECWZLghIMiUQKBNCQEiY6QQDFYGLcRyPhLYCYQJvFRQDFNIBBIIBABAEEEnwEECWYLQlKMCQSKRBAQ0qY6AQBFIOJcR+NhLcAYgJtFhcBFNMABkEpgJZFUTzXL+5fEshhH7PoHJnU/sFBVtEJoOXdx+16OtmmMhAlmC0J7RIsuX8owWxJKJdgm9lT/d+HVwLIXifsB1B6/4g/EtrGuI9GQjmA1tN6+rydP7xSgo0cjSOT3j8YyCqKM6CS+YQAGjl6M6BtQv9gIKsorQHtbLOeFpRg1jphvwRL7x95N02M+2gk1C/DL5trGOW5jACy1gkHAZTcPwSQLQnuA4JEOLkPCAMZhQACEQQQSNAOIC7Dm+uEhxLsQLLL8NKOmBj30UhkK8GKPYuyC9d/LXq+3+rXwkAf0u/HgMT6x8CYJflapD52FvbjykOf2Emqb4fBoAQDCawBgQgCCCQoBdBm1syUOyp41oB8S7QUdP6yU/uHNSBbEkcFpRsRi92trKv9C5EcAWRMQjuAkvuHALIloRxAh1vpSys9vIrl+r8dBkP1UYxk/sFAVlF9GLVkxbNg40bzYdRtMv9gIKv4mAFRgtmSUC7B0vuHEsyWhP4a0O4UxhqQvU7YD6D0/iGAbEloB1D1GHNNx/mLEmw06ByZ1P7BQFbhPiAQwX1AIMFHAFGC2ZJQvw/oXSjBXEuol2CJ5QggWxIEEBKJFHwEEDNos7gIIAxkFgIIRBBAIMFHAFGC2ZKgBEMikQIBNKSEiU4QQDGYGPfRSHgLIGbQZnERQBjILAQQiCCAQIKPAKIEsyVBCYZEIgUCaEgJE50ggGIwMe6jkfAWQMygzeIigDCQWQggEEEAgQQfAUQJZkuCEgyJRAoE0JASJjpBAMVgYtxHI+EtgJhBm8VFAGEgsxBAIIIAAgk+AogSzJYEJRgSiRQIoCElTHSCAIrBxLiPRsJbADGDNouLAMJAZiGAQAQBBBJ8BBAlmC0JSjAkEikQQENKmOgEARSDiXEfjYS3AGIGbRYXAYSBzEIAgQgCCCT4CCBKMFsSlGBIJFIggIaUMNEJAigGE+M+GglvAcQM2iwuAmhbYCCjEEAgwkkA6fQCxPgIIEowWxLuSrCR+Gc0Et5KsLEYyEQnCKAYTIz7aCS8BRBTaLO4CCD8YxYCCEQQQCDBRwCNZQptohOUYDGYGPfRSHgrwcZiIBOdIIBiMDHuo5HwFkBMoc3iIoDwj1mUAmgzK2ruX5LIcR+ZWXSOTGL/EEBm0QmgZfHUvFjtX4jkbEwbKcFCCip/2sn9QwlmS0K5BNvMDrZZPryK5YwMGgEUUtAIoPT+IYBsSSgH0Hr6vH+56phE9yzBer0bBkTj0CT3DwYyi48ZEJjFxQwIu5lFaw1odwpjDcheJ+yXYOn9QwlmS0L9Mvx62lzF6Dh/EUC+JdTvA0rtHwLIloS7+4DAKtwHBBIIIBBBAIEE7QBaJrmKYWLaSAkWUlD+007kH0owWxLZSrBiz2Lx/feLxbXfF9vm+9b59zHuRxJfaPunsDFu4u8Luc7C2H5cfegTOymtHOTCRwkGVmENCEQQQCDBx8OoJupW1oBCCi4eRjUxaEiEFHgYdUgJE51wEED4Z+QSygHEoxi3g49HMcAqPh5GBbP4eBgVrOJjBmRi2kgJFlJwMQMyMWhIhBR4GHVICROdsB9A+GfsEtoBlPphQjCLi4dRwSzcBwQiuA8IJPgIIBPTRkqwkIKLADIxaEiEFLIFUC8W/d5uVsJEJxLvR1pf4J+bkGgpXH3ENe00wMZNSJjohI39GBQTg4aESIEAsvGHa0KCAEJiaAUCyMYfrgkJAgiJoRUIIBt/uCYkCCAkhlYggGz84ZqQIICQGFqBALLxh2tCggBCYmgFAsjGH64JCQIIiaEVCCAbf7gmJAggJIZWIIBs/OGakCCAkBhawZvnAGBEEEAAkA0CCACyQQABQDYIIADIBgEEANkggAAgGwQQAGSDAAKAbBBAAJANAggAskEAAUA2CCAAyEa+AFoVxd1HkcJmVhRd/4/fHsw7/z+dV/HpsSgmsh4sy/14FvXhL16qb4IhbRQSjeggjMU/N26gbAG0Kvd0JXLQZlY2X0qPXTnoIgOtytbrqagTy2ooJAZaT++roy8Y0kYh0YgOwlj8c+sGyhVAm1mVk/OJQOLTYzXky3ro4llPRQZq9kPUic1sshUNRXneao5+9JDuFNKM6CCMxT83b6BcAZTM67KzYNmFh19IDPTpJ7LNb+X+WRVPK1l87BV2/xKO6CCMxT83b6BsAVSP+0puoLlMouyHqIZf3f92Kl1IEM+gm3EUDWmrmXBEh2Es/rl5A+UKoCYl5SfblezYVbNOkYGW1dSzOQXFI15PrY++aEiP/hGO6ECMxT83byDnAbQSriEuS/PIAuhOfiauzhifHiV/CAn9Ix3RgRiLf27eQL5LMOnZuu6FLIDqPWiq59hOyJcz0s2gfcx/xuOfmzeQ60XopfSvpbp/ohDdQ9GMu2glMcHJXLiGuD34RzyiQzEW/9y8gRxfhi93VnTz1R7RGWw9rfogOhM3x10ksRJeRd1vPtGIDsBY/HPzBnJ8I6Ks6j0inEKXrZsjF9+BNCW8aEh3p0An85/tePxz6wbK9yjGUrpyv5v/ii+ECA20kt/PP5dKHOa/0aNRK6Qa0UEYi39u3EA8jAoA2SCAACAbBBAAZIMAAoBsEEAAkG8apgAAAAFySURBVA0CCACyQQABQDYIIADIBgEEANkggAAgGwQQAGSDAAKAbBBAAJANAggAskEAAUA2CCAAyAYBBADZIIAAIBsEEABkgwACgGwQQACQDQIIALJBAAFANgggAMgGAQQA2SCAACAbBBAAZIMAAoBsEEAAkA0CCACyQQD1Y3n/sl1Pn3J3A5yCf84ggPqxmT1tlw+vubsBTsE/ZxBAPVnd//avP+buBLgF/5xCAPVlXkxydwEcg39OIID6siqec3cBHIN/TiCAerKZ/f39S+5OgFvwzykEUE+WD/834yIGxIJ/TiGA+rGePm9XnMIgEvxzBgHUj/nDazmLnuTuBjgF/5xBAAFANgggAMgGAQQA2SCAACAbBBAAZIMAAoBsEEAAkA0CCACyQQABQDYIIADIBgEEANkggAAgGwQQAGSDAAKAbBBAAJANAggAskEAAUA2CCAAyAYBBADZIIAAIBsEEABkgwACgGwQQACQDQIIALLx/5xLRErs4GH9AAAAAElFTkSuQmCC" /><!-- --></p>
<p>Using functions from the R package <code>dplyr</code> [<a href="https://CRAN.R-project.org/package=dplyr">12</a>], we can extract information such as how many pairs are mutated in how many patients.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">res_pairs <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(num_patients) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(num_patients) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">tally</span>()
<span class="co">#&gt; # A tibble: 2 x 2</span>
<span class="co">#&gt;   num_patients     n</span>
<span class="co">#&gt;          &lt;int&gt; &lt;int&gt;</span>
<span class="co">#&gt; 1            2    14</span>
<span class="co">#&gt; 2            3     2</span></code></pre></div>
<p>We observe that most pairs are just mutated in two patients, and a few are mutated in three patients. These pairs in the results tibble were those mutated in more than one patient and who tend to be clonally exclusive.</p>
<p>The average rates of clonal exclusivity can also be visualized with the function <code>plot_rates_clon_excl</code>, which makes use of functionalities from the <code>ggplot2</code> R package [<a href="https://cran.r-project.org/web/packages/ggplot2/index.html">13</a>].</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot_rates_clon_excl</span>(avg_rates_m, clone_tbl_all_pats_all_trees)
<span class="co">#&gt; There are rates from 82 patients.</span>
<span class="co">#&gt; The average rate of clonal exclusivity is between 0-0.68</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAeAAAAPACAMAAADaOVq4AAACwVBMVEUAAAAAADoAAGYAAIsAOjoAOmYAOpAAZrYbD5AdEZEgE5IiFZIlF5MnGZQpG5UsHZUuH5YvIJYxIpcyJJgzMzM0Jpk2J5k4KZo5K5s6AAA6ADo6AGY6OgA6Ojo6OmY6OpA6ZmY6ZpA6ZrY6kLY6kNs7LZw9L5w+MJ1AMZ5AMp5DNaBENqBGOKFHOaJJO6NKPaRMP6RMP6VNQKVNTU1NTW5NTY5NbqtNjshPQqZQRKdRRqdRRqhTR6hUSalVSqlWTKpXTatZT6xZUKxaUaxcU61cU65dVK5eVq9gWK9hWbBhWrFiW7FkXrJlX7NmAABmADpmOgBmOjpmOmZmYbRmZjpmZmZmkJBmkNtmtrZmtttmtv9nYrRoZLVpZrZraLdsabhta7luTU1uTW5uTY5ubblubo5ubqtuq8huq+Rwb7pxcLtycrxzc7x0db10dr11dr52eL93eb93er94fMB5fcF6f8J8gcJ9g8N9hMN+hMSAh8WBicaCi8aCi8eDjMeEjsiFkMmGksmGksqHlMqIlsuJl8yKmcyKmc2Lm82Mnc6Nns6OTU2OTW6OTY6Obk2ObquOoM+OyP+PodCQOgCQOjqQZgCQZjqQZmaQkDqQkGaQkLaQo9GQtpCQttuQ27aQ29uQ2/+RpNGRpdGSp9KTqNOUqtSWrNSXrtWXr9aYsNaZs9eatNibttmcuNmdutqevNufvtygwNyhwt2iw96jxd+kx9+lyeCmyuCnzOGnzeKozuKp0OOq0eOrbk2rbm6rbo6rjk2ryKur0+Sr5OSr5P+s1eWt1+at2Oa2ZgC2Zjq2kDq2tpC2ttu225C229u22/+2/9u2///Ijk3I///bkDrbkGbbtmbbtpDb27bb29vb/7bb/9vb///kq27k///r6+v/tmb/yI7/25D/27b/29v/5Kv//7b//8j//9v//+T///+T5k9/AAAACXBIWXMAAA7DAAAOwwHHb6hkAAAgAElEQVR4nO2diZ8k53nXexdJE+5wM5IxiRluA4YZWeYIh8EJBBgvhIRkuDIrlB1zm/sQOAZzhFlpMyxHDLYREG4IV7jZlYQWMiQYsa2JDbZGTI8zR9dfQb1vHW9V9VP1Pl1dT731dP2+n480O11T1dXvt986fvUekwisNZPQOwBkgeA1B4LXHAhecyB4zYHgNQeC1xwIXnOWFnz87k+n//r8b9hq9Zbnv/Ex6+/+x2Sy8Xbxhev7T71d98fcv6j9m9LL9pfF/eRsftm/N+9S+3fcompgecGTG68kb745aSf45CZrr88mseDSKz0LXtxPCcHmXWr/jllUTSwv+Ma7Eq8nT2wKC96rvCIqeJHVS5crWHAXWgi++R12r6/vf4MV/KVbk8m73zRLPveuyeTGb7Yf6/veZf9lub6/cTK5+Wa2+Pp+XDO3Cutd//X49W/Kt29ef/JNc6SYTLKP97nNyZOvZOWV/UHxbQpv/XZhO2b7J+bNzsz/CltJ/i7eM/f25qWkPC93N+Jfvmj3M/mWXe5upR/Frlr8dFQBFN/qC7fc38abOUsWROXScB/s3cUPlhXVSrQQ/NmnzS6eP/NZIzg+UMeYo/bJxLKX7Jf9l+X6/hObk7jw0hfTvc7XS/86q61n2esFwWf5n8blcFZZ0axZeOtMcGH7Nx9f7sZbKm7FCc7f3rx0vrln324vFxzLNq+kZ6Vk1dKnowqg8FZP3ir87eXuV2xW/rYguFoieVGtRgvB//u+edPjjfNY8PV98+3//KYpqbiWxntpyyy+ODrJTqDxr1tRYXF61snWO998z9vJ6+kff5P5+Mak+4bEn/pzk62sfNM/cG9TfOtUsNt+vJ34CLJX3UomOH979/LxzcfuHHxs/5dvtfi2lTcq7kX+VuaDfH/2t5e78arX31va4/wcXNxQ/g5hDtGPzee93N0zgs+T83CyI//3+/7GZrL3j+2BLiuCdC/TxfaP3Xrnm098w//Kt56aPo5LKBd8np3rk0qW/UHpbfK3TlUU9+vYXqyVtlIUnL594jR+X7NFJ9jsRXaETlctfTqqAApvVfrby127d8fFv80Fuw0V1gok2ByxzoyareRglBxZkkNMXsr2BJd8Svux8sV2rwvrmWPx5CvfTP44OUjaY2JB8J7bkvuDwttU3jqKituPlyb/L2ylsG729ulhcsu+rxNsyjk7QldXrbxRvheltyr8bWr6pPi3uWC3ocJagQSbb/TxRlQWHB9/vuJv/s8v7NYIdourgqPP33Kn21aCq28dlQWf2ZNgneDs7e1L5n/H2X1LUrrpEYtctfRGbi/8giulMTTB5kIr/gzJIdpdbOxFxRNhRbBbnB6iS/dA/+evpTaXOkS7w2z5raOouP3L3SdvuZNCuj/HpWOneftk1ZMbnzEvFgSfTb4xO0JTgqkCKL9VUXDyqzn4lUqj8M0trxVK8NmN35QWWXxp8On4At/uc3xlEF/pLxRBJjhbbA9Rbr2zyU+Ja8735tepNRdZ57vJhosXWU5w+a2TddLtm7c7sxc8ha3EB2Z7FxIffrO3z3bz683b5idk88+flB2hKcF0ARTfqig4fjN7kbVQGnaldENFwflbt6aVYHM5mNas9MiylV/c1wh2i80twtbCetlRML8LKgQdZ+ntR/U2qXKrUxTstp9U7PyskG3F/PvJWxvu7fPdNJUmEZzco5zkt+OUYKoAym9VFPzEZnI2qpSG26VJepmfCQ5ym2QvTPeyQ+d5fAp7wuQUNoD49HHpG5iVSnFxdHnL3Abk69mk4T154nOe5hjFJOtzm3kW4f6g8DaVt863Y7Z/nFxhbVS28v2bk6/8Yh50vOftbDeT+x/7S7Kf7haOFkwUQOWtiufg+OtpryfLpfHF/IM9UV4r3YVVwNMkH4uRaVvyc36fQLCHL91a7gFDAxA8POKrjZXPgoWNQfDQiK+G3tPZxiAYdA8ErzkQvOZA8JoDwWsOBK85ELzmLCv4/+UU/knQvHSllWU3LVLKAWkv+MUXw1kQ3LRIKQcEgitLRUo5IE2Crw633/+O+cfF829lr7nCgGAVNAiePziKTp+L/zHbfhaCtdIg+Oqlt2zdne68jhqslgbBF9/yTnR191GUH6J/bIxb/OKLwrsGuqBB8Oz9FcEG921HDVbBMjXY4AoDglXAOAdDsGYar6IPkqtoCFZM+/tgRJUqaJ9k/aKWdGFBamUIhmBtMA7RF/vb20fZa64wIFgF/qjS3Cld3H6UvugKA4JV4L9NmpkL6WlWhV1hQLAKWEFH+qMcVbYVLPyBQBlWVGnuiFPctx01WAWcGnx1mPuFYG0wosqL/SP3oisMCFaBP6os+YVgbfjvg0+3DYtX0YgqVYBms5WlIqUcEAiuLBUp5YCwniaZTCtlGBYENy1e4j3DaVUZn4YhWCucFh0XL7wMwVphBB3zh28kh+hyVAlUwIgqTw9wDtaLvwbHPyBYL/5zcBJ0ZGn0MCwIbrqXUu8RVqtK1GC9tL8Pfp8cENwd7ZMsCFYBowabk3DefxSClcFIsqb040IIVoH/Knr+8FHhRQhWhv8+OD5Qp4+Dy0mWpGDQGf4ky7SJdrUYNVgZvGazVLtoCFYBq38wBOvFn2SZI/X8VdwmKYV3H7yTX0h3U5CIKnujfZL1MUfvFgQ3LVLKAYHgylKRUg4I4xA9f0AeoiFYBbyocpY8VYogWB3+2yTzw+EKA4JVwGmy81p6iC5HlQXB0jsJ2sOIKvePrOsE921HDVYBq9FdnlhCsDoY5+B7EKwYRqO7KQ7RimHcB8c/qBHfEVWqoH2S9YGAFgQ3LVLKAYHgylKRUg5I+yEcIFgFrP7BZFQJwSpgtehwd0kQrA1Wm6y0Hpejyg/I7hjoBs5QhoUKjBqsDU4NdmdgCFYH5xw8dUNVQrA2GFFlqe+KKwwIVgEnqrxLCkZUqYLVGt2FsiC4aZFSDggEV5aKlHJAeLOuUE+TIFgFvFlXTmtaVYayILjpXkq9R/y3SfZ2+KXFvkkQrAJGB/C8Bi+2qhTfPbAqjKgyH03J4L7tqMEqYLSqvP0omhGj7ECwChgzn72/vlVlKAuCm+6l1HuEMetKTQ1GkqUCxn3wjO4A/l2rAMF90T7JgmAVtE+yIFgFjCTr8IhMsiBYBcwki7iKhmAVLNO7sJxkrSYY9IQ/ybKH6B3UYKUwGt3FF1l3iLEqIVgFvKEMqadJEKwCf5JlnyY9l73oCgOCVcBLsqinSYgqVdA+ySJn9u7JguCmRUo5IBBcWSpSygFpFmyvsWoe+EOwChoFz0wIXegmHEGwOpoET3dej2twceB3CFaH/xBdF1VWBcvuJ2iJX3Bdkx3UYBUsU4MNrjAgWAV+wTgHq8Yv2E0jbHCFAcEqaH8fjKhSBSsnWUEsCG5apJQDAsGVpSKlHBDGIToqXGNBsDb8UWX+w+IKA4JV4I8qsx8JrjAgWAXLHKLJqFJw30AH4BxcWSpc3r0DwZWlwuXdOxBcWSpc3r0DwZWlwuXdO+0FI6pUQfsk630BLQhuWqSUAwLBlaUipRwQxiH6Yt9NugLB2vBHlbb/2e3FFh0QrAJ/VDkzD/uni/MmQbAKWFfRZKvK9wnuFegMlmDTaifFfdtRg1XAEXx16GblcIUBwSpgXUUfuVdcYUCwChjtoot+kWRpwy+4bvbRjzUAwYNhtUk5IHjwMM7Bs+3aSTkgePD4kyx7mCZ6NkCwChiN7qKI7JsEwSrgJVlJDV6clKMOuf0FS8IRfLFPDQiOGqyCZbJogysMCFYBTzD1NAmCVdB+CAcIVgGjBp/Sk3IgqlRB+yTr6yh6siC4aZFSDggEV5aKlHJAmA/8iUZ3EKwCTv/g+CwMwVrhRJUXL7wMwVphHKLnD99IDtHlqJIULLqvoAWc26QDnIP1whrKEIL1wm2yk7WrdIUBwSrAbVJlqXB59057wYgqVdA+yYJgFUBwZalIKQeE9zTJNaschgXBTcsXeb8wosopejYoxh9VzvOZRw3DsCC4aekC7xvGkP6HWc+VclQJVMBIsm4/KtTiYVQzwU3LF3m/tG90B8EqgODKUtni7h9eq8r5q7hNUgpaVVaWipd4zyDJqiwVKeWAQHBlqUgpB4TTZOcBDtF64UWVs3zqs2FYENx0D2XeK7wJoh3DsCC4aekC7xtOm6zX0kM0okqF8MbJMnMIJwyjmgluWr7I+6X9BNEQrALG06R7EKwYxm3SFIdoxfAmiKYGQiObza4CBEvQcf9gCB4a7QcjhWAVsPoHk0kWBKuANZShu4iGYG2wWnSkY5Ey+gevgtyHHDOsORtcBUYN1gZHsDsDQ7A6OIKnbtIVCNYGa4yOQtcGCFYGJ8m6SwpGVKkCtMmqLBUp5YBAcGWpSCkHhDd/MPWwAYJVwJs/+BSN7rTijypti46X0HVFKYwWHXkNRqM7hfAe+LsoaxjVTHDT4iXeM7wO4DMMwqKV9pNyQLAKUIMrS+WLvF94s4+i85la2idZH/IhaEFw0yKlHBAIriwVKeWAtI8qIVgFjKjy8IiMKiFYBcyokrhNgmAVLNO7sBxVegXL7jjgwRmrMj5E76AGK4V3kXWHGKsSglXAnAGceFwIwSpgPi58LnvFFQYEq4AXVeJxoVraJ1kvZgSwILhpkVIOCARXloqUckCaBMeXz6bjd02LDghWQYNgc3V1cfuRmfaMusiCYBU0CJ4Zq9Mjc4uUj/wOwdrwnIPjWlwXVeaCRfcPrEiz4PmDg9o2WajBKmgUfHV4ENU+TYJgFTRfRZvBk3AOVk2D4MSvPUxTV9FIslTQIDgbAa3mPvir2HRvQWrlcQkmcYUBwSpgJFmFCdAgWB3+JKswomEEwerwJ1luREODKwwIVoE/yYryQ3Q5yeILFtx94MOfZEU4B2vGn2RFEKwZf5IVQbBm/ElWBMGaYSRZNYIRVaqgfZL18Y9/PJgFwU2LlHJAILiyVKSUA8KIKvPE0uAKA4JV4I8q88TS4goDglXgjyqzxDLBFQYEq4AVVZKN7mLBgvsFOoIVVaY/DO7bjhqsAk5UmSWWBlcYEKwCRlTpEq0IgtXhjypLfiFYG/6osm72UUSVKmifZH1ZGCB4OSB4xILTjHK2TTa6g2AV+KNKOws40bMBglXAaFUZRWTfJAhWAS+qTGpwOaoMJFiuKNYTTlR5sU+N+B5IMGrwcrBaVZL9gyFYBaxWleTjQghWgT+qrBvCAYJVwGhVeYpZVxTTPsn6jwGBYDYQPGLBeXNKM9hdCgQrg9MBPD4LQ7BWOFHlxQsvQ7BWGFHl/OEbySG6HFWGFCxZImsGI6o8PcA5WC+soQwhWC/+qDLJO7KGsxCsDFYHcLIGI8lSAasDOCn4L/QDBK9G+yQLglXASLJMTX52sckOBKuAkWRN6Z4NEKwCf5I1zyemNECwMvxJ1tVh1nOlnGT1JFjyw48Bf5JljtJzYnrZngSjBq8Gr9Ed1SYLglXQvtEdBKuA1+hu/ipuk5SCRneVpX0Uep+0T7I+YgljQXDTIqUcEAiuLBUp5YAwosr5A/IQDcEq4EWVs3xmLFcYEKwC3vzBDlcYEKwCzvzBr6WH6HJUmQiW3DfQAYyoMr4dNlPMJrhvO2qwCtrPHwzBKmDMH3wPghXDaHQ3xSFaMbz5g6lxshBVqqB9kgXBKoDgylKRUg6IP6rEaLOqYfUPJqNKCFYBayhDd5cEwdpgDWWYjkVajiqBCjhDGRYqMGqwNjitKt0ZGILVwWlVOXWz6kCwNhhRZanvyjAsCG5auLx7hxNV3oVgvbRPsj6aEcCC4KZFSjkgEFxZKlLKAeFNEE09TYJgFfAmiD4lokoIVoE/qrRNdl5a7JsEwSpgdADPa3A5qswFi+4fWBFGVHl1uO2iLPdtRw1WAaNV5e1H0YwYZQeCVeCPKusm5YBgFTAmiK6pwUiyVMCIKmd0B/CPqASCPUCwMtonWaFVtQOCHVmSdXhEJlmhVbUDgh3FJKu2b5I2ILiM7R+cCab6B2tDtDCHiD/JsofoHdRgpTAa3cUXWXeIsSpDq2oHBBdwQxlST5NCq2oHBDtSv/Zp0nPZixCsDF6SRT1NQlSpAnQfrSwVKeWAQHBlqUgpB6RJ8CyZbaXmgT8Eq6DpIuv5t8zVlZkVi7rIgmAVeA7RsWRzi2RcJwzDguCmxUu8ZzyC46pbF1UCFTQKvtjfeVTbZAc1WAXLPGwwDMOC4KbFS7xnfLdJyXDCOAerpel5cHJsNk+UcBWtlqYanE63gvtgzbRPsr4toAXBTYuUckAguLJUpJQDwogqo8I1FgRrwx9VGtFUs1kIVoE/qoymO6+jBqvFH1VG+SG6HFV+m9xOge7wR5URzsGaYQ1GCsF68UeVEQRrxh9VRhCsGUZUWSMYUaUK2idZX1ulRwuCmxYp5YBAcGWpSCkHhBFVpv3AE1xhQLAK/FFlYfKVCILV4Y8q3eQrBlcYEKwCVlRJtqpcECy2j2AFWFFlMvmKxX3bUYNVwIkq08lXLK4wIFgFjKjS9QOPIFgd/qiy5BdJljb8UWXd7KMfXBkRRxBcpn2SBcEqYCRZeds7gysMCFaBP8nK295ZXGFAsAoYje6iiOybBMEq4CVZyY9ykrW6YLEPBRycJCsLtAzu244arAJWozuyfzAEq4DV6I58mgTBKvAnWXVDOECwChiN7k7pSTkQVaqgfZL1XkMgC4KbFinlgEBwZalIKQeEEVWapw3ExFgQrAJG/+Ap/bgQglXgjyrn+XD+BlcYEKwCf1R5dZg9Di5HlVaw5K6BLvBHlaZN9JyYlAM1WAW8qJJKsiBYBe2jSghWAS+qnL+K2ySlIKqsLO2hzHulfZL1c5eiWwtSK0MwBGuDEVXOH5CHaAhWAS+qnBETREOwCnizrjhcYUCwCjizrryWHqLLUeVygiU/A2iAEVXuH9mpdRLctx01WAXtZ12BYBUwZl25B8GKYQxlOMUhWjG8WVeo3oWIKlXQPsn6qgYgeDBAcGWpSCkHxB9V1g3hAMEq4My6EpFRJQSrgNUB3N0lQbA2WB3A03pcjiqbBIvtLlgWTgfwQgVGDdYGp1WlOwNDsDo4rSqnbqhKCNYGI6os9V0ZhgXBTcsXeb9wosq7pODvDkUowSeTrc6KPeX86Vc8f3EyuVH8E/8KVdonWWMTfH3/628+XrK0fHh9Xe7uLbfCArxJOaiHDWMTfHbzs5t7tUvb4fVV/YNOBRcm5TglkqyxCT5+6ov3N+J6bI7TJzcfX9+fTOIaff7Md8Q/zjcnE3MAv9yd3PjUM4+jdGGi5FPxwr1ETfy/5Pet8/TFT04mT70dRaWtmdXMCxuR+auNyP1ut5L/M9lw/mZ2Jxa+gv4ky7boeGmx68rIBF/ubhmv0UnsI7Z8Hcs2/z7f3EgPpPFC8zeXu0Z+stAK3ox/xstywfb3Sfpi/Mu1/doUtpb4TF7Maqz5t3mX+Pd8UbrhfF37BgsHGUb/4LwGl5OscIKDcGYq2GZSE+P/zK+myG2J/oBRmb0Yl3m20Kxn/8DW3EzwXuHF9Gdpa9m7mf9lggs/80XVdZ8hrxAYSdbV4bZLOkZbg483snq1ZerLycSylZb8WfzvG6/YShsXdLYwV1MUnP5MfnkmkVPZWrw9u6X0rzPhqeDiovh/+ZsdZ4dzvuCk0d3tR9GMGIRlXILjs6shLuez+Fy8lR2AE1WXuzdstc4Fpwv5gotbW1awe7N4Jxcu9P1JVt1Id+MSfGJLzhwWLz9sLqPObhSOnbbM41eyo+dZ4d61WXD6s7Q1K9S8UHOILi4qrJsqXuIcnE3KgRqcHJyzH8cTe7SOpcZlmwg2FXjzxivuIitZWBRsll3fv1ERnF9kFbaWvVH1Iiv9vbTI/p6sa79di7dRjCRrhv7B+dXpSWztzN6LmJuT5MAcGeeTG5+JK4+5TfrkzcfZwqJgexfzjR+u1uBPJifO0taiqHRbVP29uCj73dbqyaQce3kFU7jC+H2GQBYEN71keSxyVjgBDwAIrixdpTDNUTk9mA+G9lElBC9iblmG5ZcTVR4ekVElBKuAGVUSt0kQrAJO/+B3yPmDrWDRfRsSX0YTerf8+KNKe4jeGXsNXk/BSWuO+CLrDjFWJQSvg+B0DEPqcSEEaxecRpX2cSExdyEEaxdciCqpx4WjiiqjH0zTrQwJ2idZ4xL8Q2hWLP0egODK0prP/UNpVi1+eZoFzx8c1bboGJfgH0bTpQoZmgWfbh9ZydRF1rgE/3CaLlXI0Bx0vPDykb1FoiaIHpfgH0HTrQwJmgTPH77x4Kg2qhwXP5Im9G75abxNOjCH57o2WeOqwT+KpmMbAjQ9LvyWd+alGmwYhgXBTdcUxo+m6VpH9zQITobXOcA52PBjaLrW0T3+26T5gwNcRUc/jqZLFTLgPriytKYkfjxNlypkaJ9kvW9ZOrQgtXK94J9As2rxywPBlaU1n/sn0qxa/PIwDtFR4RprtIJ/Kk13IqTwR5XmeSHVbHZcgn8aTYcmhPBHldF053XU4J9O06kLEfxRZZQfostR5dKCRT+HND+TJvRu+SkJvtzdK3YxTaLKCOfgmJ9FI+KkU0qCjzdMP9iTtPNFGlVGEBzzs2mErHRIUXBcgU3Xqaw3eRpVRhAc814aOTFdURFsuimfuWEAmmrwuJKsn0MjI6VLioKv72+ZDpDHrn9ck+AP0vRiQXDTNeVUM1uQiJNOKZ2DzzcnG9FxYwdmVxjjEvzzaLoTIQUjybrYd3NyjFbwz6fpUoUMlXOw+VE4B5sky3ZPur34wH9cgn8BjZCVDmkUbJOsmXkWPF2cVmdcgj9Aky2+vt/1OKVd4QSnI6YVBiHIkyyy0V2N4P52vVd+IU22+GRxFNCBQNTgnDzJMo06Uty3fVw1+BfTpEvP331Lg+AKeZJ1degmbXCFMS7Bhl9SorDg+q/+bQWH6CgdcXgyqSRZZgrwHFcY4xL8S2mShSdbGs7BETHEk202W/Q7WsG/jMYuO//Jb+sQvDiSpe2XVDM55biiyl9OY5eduMGDh0e5Bvu/hq4wampwY73uyoLUyvWCfwXNMkUXhtI5+Mw/rYgrjHEJ/pU02WIdgtNBryeVp0npiIYJrjDGJfhX0Uhp6Q5/o7vCNMLReAV/NU2XKmRgNLqLIrJv0rgEfw1NhyaEKAuOrwf33BwALqpMajArqqTp5bNI8qtpQu+Wn3KbrKe+kLTaSciiynQaYYv7to+rBv8aGiErHbLQZGcvv5Z2je7I/sHjEvxracS8dEaDYNfobpnHhWsq+NfRiHnpjNIh+sQcok27u4ymIRzGJfjX0whZ6ZDyRdZZJXOzh+hTzLoSRV9HI6GkW9p3H61p47AsXTsSEvxbaFYtfnkguLK05nP/VppVi1+eXHDSL4mKKvMfBlcY4xL822jEvHQGp39w9sPgCmNcgn87TXcipPC3qiwkltF4Bf8OGhkpXdIkOI0q88SyHFV2JLi3T7oa30wTerf8NDWbTaPKvHGlwX3bOxLcdSUUqsG/k0bQTEc0NNlxQxlCcPQRGgkl3eIfyrCYWI5X8O+i6diGAA3NZiPcJjl+D42Ik06p9A++vr9XPFA3CR5XVPl7aUScdEr1HHy81TzDsSuMf7ROeAX/fpouVchQFXyy0dy2cqyC/wBNlypkqI6yE9t1bXbyp0muWeVYBf9BGikt3VESHJ+Eo+PJjVfyF2xGmT/sN4xV8B+iETDSMf5WlfN85lHDWAX/YZruREjhH8rw6jDrmlSOKkM76RRvOf0RmlWLX56C4OPiHbAhySjN+BxzYv7g0E46xVuD/yiNjJQucYLN8EknRcPFjJJodBfaSad4Bf8xGhEnnZILtt2nSn2oPK0qQzvpFK/gP0EjZKVDii06InsZXSRrVTl/dezT6vxJGiErHeIXXNeqsmZwP4O4BcFN15TTn6JJFp5Vr14GRLPgRVxhjEvwn6axy86feRydVEe/GAqMGjx/gBoc/RmafLmRPEgKgvMWHeWhDM0F1iyfGcsVxrgEGz5Worxs+DWYIp8/2OEKY1yCv5UmXXq+Wch3hwVn/uDX0kN0OclqECy8zyH4szT58ssPD9Swf/5gM06WmWI2wX3bx1WD/xyN+4PjgY7C0n7+4HEJ/gSNXWZaSGiswdn8wfcgOIr+PE2y8GQyUXkOTm+TpjhER9FfpOnUhQi8+YOpcbLGFVX+JZouVcjQvvvoP1uOTi1IrVwv+C/TrFj6PQDBlaU1n/vbaVYtfnn8h+i60WbHJfiv0HTqQgRW/2AyqhyX4L9F050IKVhDGbq7pNEK/rs0HZoQgjV/cDoWaTmqXFKw5Ifogb9HE3q3/PijylIFHm0N/k6aLlXIwJl1xZ2Bxyv479N06kIEf1QZRVM3q85oBf8Dmk5diMBIskpdG1xhjEvwP6TpToQUnKjyLil4XFHlP6bpToQU7ZOsr12Gji1IrVwv+J/QrFj6PQDBlaU1n/uf0qxa/PLwJoimniaNS3DNlUWXKmTwR5XmFHxKRJXjEvzPaTp1IYI/qrRNdl5a7LoyLsHfRdOtDAkY/YPzGlyOKpcSLPwp5PkXNKF3yw8jqrw63HZRlvu2j6sG/0uabmVIwGhVeftRNCMGYRmX4H9F07WO7vFHlXWTcoxL8L+m6VpH9/hvk1CDDf+GplMXIjDug2eYdSWKvpumSxUytE+yaqaK8tGJBamV6wX/W5pVi18eCK4srfnc/55m1eKXp31UOS7B/4GmSxUyMKLKwyMyqhyX4P9E06kLEZhRJXGbNC7B/5mmWxkScDqAZ4LLUWVLwdIfSIj/QhN6t/z4o0p7iN4Zew3+rzTJQjMVwkD7f3NaVcYXWXeIsSrHJfi/0dhlpvP3+dMD7SDMaVUZkY8LxyX4v9PYZWdmhB2FQzhEyW2SfVz4XPbKMCwIbrqhNL6nRHmZxiEcIhdVUo8LP7o03cQDrnAAABDlSURBVFmQWrle8PfQZIv9QwSGon2SBcFO8OXuUP02X0Uns3HUPPAfl+BGzjcHegKOmgVP81mxqHMwBGcM2W+T4LTLirmCvnh+8SoagjOSeVsHKrlBcDobR12Stbxg8Q8DFmkKOpLZOOqa7KAGq8B3FT09qnvYAMEq8AvGOVg1DYLT2TjmDw5wFa0Xz32weYpUcx88vqhSJe2TrA/V0YMFwU2LlHJAILiyVKSUA8KIKi/23UiGEKwNf1RpbpHMLXGCKwwIVoE/qpyZK2hi7kIIVoE/qrT/JKLKWsGyOwyWwx9VRuaBUj4Umvu2owarwJ9kmarshrpzhQHBKmAINhMn5bjCgGAV+KPKkl8I1oY/qqwb0h9RpQo67T7amwXBTYuUckAguLJUpJQDwogq0+bRCa4wIFgF/qgycpOvRBCsDn9U6SZfMbjCgGAVMKLKfPIVb/9g6Z0Fy8OIKrPJVyzu244arAJWq0oI1otfcKGbcATB6vBHlRFukzTDaFVZIxhRpQraJ1k1PWZ7sSC4aZFSDggEV5aKlHJAGFFlnlgaXGFAsAoYUeWUfh4MwSrwR5WlqQshWBv+qNI1rixHlc3DkoCB4I8qXeNKg/u2owargNOqMiIbvkOwCiC4slS0tAPgjypdYmkYhgXBTfdS6j3CiCpP6VlXPtAF3TuC4DLtkywIVgEjyZo/QA3WCy/JmhGTckCwCnhDGTpcYUCwCjhDGb6WHqLLSVYngqU/HmAkWfvJeJUJ7tveieDuKyFqcJn2QxlCsAoYQxneg2DFMBrdTXGIVgxvKENqckpElSpon2T9O0MgC4KbFinlgEBwZalIKQfEH1XWDeEAwSpg9Q8mo0oIVgGnf3DhLgmCtcEayjAd8L0cVVrBovsGOoAzlGGhAqMGa4PTJsudgSFYHRzBUzdUJQRrgxFVlro2uMKAYBVwosq7pGBElSpon2RBsAoguLJUpJQDwpt1BU+T1MKYdeXwKDolokoIVoE/qqxrsgPBKmg/QTRQgT+qtIfoHdRgpbAm5di+Q3QAh2AV8PoHu94Nw7AguGnZ4u4ff1Rpzr/UBNEQrAJGVDnbxgTRelmx0d0AgOBGIHjEgtOe3/H9MHWIDi02A4Ib8USV8YWWGUuYusgKLTYDghtpSrKSeyPz4+L5xduk0GIzILiRpiQr6fldF1WGFpshXkS6aRKc9Pw2t8PUw4bQYjNQgxtprMHWbN3TpNBiMyC4kaZzcNLzG+dg1Xiuou2kSQeIKvXSJDjt+V1zHwzBKkCju8pSkVIOCCPJigqnYAjWhj/JMo+T0KpSLf4kK5ruvI4arBZ/khXlh2g0ulOIP8mKcA7WjD/JiiBYM/4kK4JgzfiTrAiCNcNIsiBYM+2TrG/OCGBBcNMipRwQCK4sFSnlgDCiyot9N5IhBGvDH1WaC2nTDy3BFQYEq8AfVc7Ms2Bi7kIIVgErqiQb3eWCZXcQrAYrqjSNOlLctx01WAWcqPLq0A115woDglXAiCpNTc5xhQHBKvBHlSW/EKwNf1RZN6Q/okoVtE+yPpbTvwXBTYuUckAguLJUpJQDwogqZ9tkozsIVoE/qjTPCqmeDRCsAkaryigi+yZBsAp4UWVSg8tRpRMsuoNgNThR5cX+DjFvEmqwClitKsn+wRCsAlarSvJxIQSrwB9V1g3hAMEqYLSqTEc0TBiGBcFN91DmvYL+wZWlIqUcEAiuLBUp5YCwOoCbwe5ShmFBcNPiJd4znA7g8VkYgrXCiSovXngZgrXCiCrnD99IDtHoAK4QRlR5eoBzsF5YQxlCsF78UWXSJitrODsMC4Kb7qXUe4TVARw1WC+sDuAQrJf2SdaHAloQ3LRIKQeEkWTlgZbBFQYEq4CRZOWBlsEVBgSrgDcph8MVBgSrgDMpR9b2rpxkfUh610AX+JMs103Y4L7tqMEqaD8pBwSrgDEpxz0IVgwjyZriEK0Y3qQcGPFdLe2TrE8YAlkQ3LRIKQcEgitLRUo5IP6osm4IBwhWAavRHRlVQrAKWP2D3V0SBGuD1T847eBfjiqtYNF9Ax3A6R9cqMCowdrg9A92Z2AIVgenf/DUDVUJwdpgRJXzh+4IDcHa4ESVd0nBiCpVgO6jlaUipRwQCK4sFSnlgPBmXcHTJLXwZl05JaJKCFaBP6q0t8MvLQ5lCMEq8EeVrgajf7BCGFFlfLfkoqxhVDPBTfdR6H3CaFV5+1E0exaHaKX4o8q6ke4gWAWMWVdQgzXDiCpnGMpQMe2TrB/ERMKC1MoQDMHaaB9VQrAKGFHl4REZVUKwCphRJXGbBMEq4HQAf4ecIJorWPwjgCb8UaU9RO+gBiuF0aoyvsi68xCClcKbdYV6XAjBKvBHlfZxITF3IZIsFfCiSjwuVEv7JOu9S9C1BamVxyU46xhc88AfglXgqcGz99vRwKlzMASroFmwucAyV9DU/MEQrIJmwabm1iVZywiW/QyggUbBeedRKotGDVZBo2DbMbjuYQMEq6BRsO0YjHOwahof+NsEev7gAFfRemlMstIgGg3fFYPuo5WlIqUcEAiuLBUp5YAwosqocI0FwdrwR5XmcRI6gKvFH1VG053XUYPV4o8qo/wQjf7BCvFHlRHOwZrxR5URBGvGH1VGEKwZf1QZQbBmGFElBGsGje4qS0VKOSAQXFkqUsoBYUSVF/tu0hUI1oY/qrTdk26jRYdS/FHlzKRZ08V5kyBYBayoEq0q9cKKKk2rnRT3bUcNVgEnqrw6dLNyuMKAYBUwokrT0z/HFQYEq8AfVZb8QrA2/FFl3eyjiCpV0D7J+nI+nVuQWhmCIVgbjKhytk02uoNgFfijSvOskOq6AsEqYLSqjCKy8xkEq4AXVSY/y1HlEoLF9h544USVF/vUiO+owSpgtaokO4BDsApYrSrJx4UQrAJ/VIlpdVTDaFV5illXFNM+yfo7fQDBq8LqH2wGu0uBYGVw+gfHqiFYK5wk6+KFlyFYK4wka/7wjeQQXU6yehEs+dHHASPJOj3AOVgvrKEMIVgv/iQruZjOMi0IVgarfzBqsF5Y/YNJwUiyVNA+yfrOpenOgtTKEAzB2mBElebns4tNdiBYBYyockr3bIBgFfijyvxa2uIKA4JV4I8qrw6znivlqHJ5waIfBND4o0ozfMOcmF4WNVgFvEZ3VJssCFZB+0Z3EKwCXqO7+au4TVIKGt1VloqXeM+0T7L+eAu6siC1MgRDsDYYUeX8AXmIhmAV8KJKd7/kCgOCVcCbINrhCgOCVcCZIPq19BBdjirbCBb9KICCEVXuH9k5hBPctx01WAXtJ4iGYBUwJoi+B8GKYbSqnOIQrRjeBNGYnFItmD+4slSklAMCwZWlIqUcEH9UidFmVcPqAE5GlRCsAtZQhu4uCYK1wRrKMP2BCaIVwhnKsFCBUYO1wWlV6c7AEKwOTqtK17gSgtXBiCpLfVeGYUFw08Ll3TucqPIuBOulfZL1iYAWBDctUsoB4c0fTD1sgGAV8OYPPiWSLAhWgT/Jsi06XlrsugLBKmD0D85rcDnJ+oTsjoFuYCRZV4fbLulw33bUYBUwGt3dfhTNiEFYIFgF/iSrbs4GCFYBY/5g1GDNMJKsGfoHKwZtsipLRUo5IBBcWSpSygFpH1VCsAoYUeXhERlVQrAKmFElprbTCqd/cCYYje4U4o8q7SF6BzVYKYxGd/FF1h1irEoIVgFvKEPqcSEEq8AfVdrHhc9lrw3DguCm5Yu8X3hRJfW4EIJV0D7J+mCZXi0IblqklAMCwZWlIqUckCbBaUZZ06IDglXQIDhtjGWmPaMusiBYBQ2C0+aU5hbp4vnF2yQIVoG/BtdFlRXB4nsKWtF4m2RPvnVtslCDVdB0iE4aY9U9TYJgFTQITqsuzsGq8dfg+YMDXEXrpekcnDanrLkPRlSpgs6SrCVY3YLUyhAMwdpgRJVR4RoLgrXhDzrMuZhqNgvBKvBHldF053XUYLVwanB2iG6MKpdA+COBIv6oMsI5WDP+oCOCYM34o8oIgjXTvgYjyVIBI6qEYM2gf3BlqUgpB4SRZMU/8jk5IFgb/vtg24X0NvomKcWfZM3Ms+ApptVRCivJQv9gvbCSLNOoI2UY1Uxw0+Il3jOc++CrQzdpwzAsCG66hzLvFUaSZaYAzxmGBcFNyxd5vzCazRb9QrA2/EkWJqdUDZKsylKRUg4IBFeWipRyQBhRZXykxlCGamH0Lnz+LQzCohdGo7soIvsmQbAKeFFlUoMRVSqEE1Ve7GPEd7WwmuxgtFm9sBrd4XGhXvw1uG4IBwhWAaPR3SlmXVFM+yTrd/Pp3ILUyhAMwdpg9Q82g92luMKAYBWwgo5TqtksBKuAE1VevPAyBGuFUYPnD99IDtHlqHIJwdIfAtTDiCpPD3AO1gtrKEMI1os/qkzaZGUNZ11hQLAKWA8bUIP1wuofTApGVKmC9knWx+WA4O6A4BELTqNKc5X17GKbLAhWASPomNJdVyBYBf6ocp7PPGqAYGUwhnA4zLomlaNKScGgM/xRpbkdnhPzB0sKRg3uDF6rSqrRHQSroH2rSghWAa9V5fxVdF1RClpVVpb2UOa90j7J+mhCEAuCmxYp5YAwkqz5A7IGQ7AKeEnWLJ8ZyxUGBKuAN3+wwxUGBKuAM3/wa+khupxkpYKldxCsBiPJ2j+ylTnBfdtRg1XQfv5gCFYBY/7gexCsGEYWPcUhWjG8+YMxTpZa2idZP6MJCB4KEFxZKlLKAeHNukIdoiFYBbxZV06JqBKCVeCPKgsjGkYQrI5lanA5qmwULL7fgAmjf3A++YrBfdtRg1XAiCoLbe8gWB3+qLJupDsIVgFqcGVpL6XeI4yocoZGd4ppn2R9pEKfFgQ3LVLKAYHgylKRUg5I+6gSglXA6V14REaVEKwCZlRJ3CZBsAo4rSrfISeIrgoW31XQBn9UaQ/RO6jBSmG0yYovsu4QHcAhWAW8/sHU40IIVoG/BttTMTF3IQSrgBdVUo8LEVWqYNkky7HSNIarrBzsjVUCwWsOBK85ELzmtBcMVADBaw4ErzkQvOZA8JqzvGDXDr7UIn7ZlS/2t7ePPH9dt25pmpBlVy4N+7X+LC3YFG0STbt/tVjZJNwm7G6zblSaLnPZlUvDfq0/Sws2D5Yunn+r9K8WK89MaU+XsVR8u+J0mUuuXB72a/1ZWrBr4FFuy7PkyoblVi6sm0+X2WJlN+zXOFhasHtKXO7TsuTKkTlsHjT/de26pekyl1y5POzX+hOuBl8dLuW3/MbLCl5pr1UT6hxsa1LLNy5Pl7nkyuVhv9afFlfRB/lV9MHyV9HZKkv7Lb/dsjW4sPIUh+hm0hEs4+rQ+j44XjmphUtZcm/c+j443etnR3QhjSRrzYHgNQeC1xwIXnMgeM2B4DUHgtccCF5zhiL4+v7EslV47UtvBtud9WE4gq3a8829/KXzp18Jtjvrw8AER8cb+UsQ3AVDFHy+aY7V5v9b9tB987FddP70pzbT1/eaNgWKDEzwWezucjf2d3LzsanB1/dj4SdPvW2WnW/GP08m5n+pcuBnOIKTi6zY7Q8YnbFdI/jMqLTG0/Nz8j8cvNkMR/CWsZccp89i0zes4JPCtbXV6v4HWAxKsD1CxxX2xitZDU6PzhYIbsOwBEfH8SH5zEg9S2pw/CP/Ewhuw8AEn29uWKnnmzdeMafe6/ux7dQyBLdhYILjk+5WdByfgT8T2z2ebNiLr7QWQ3AbhiIYCAHBaw4ErzkQvOZA8JoDwWsOBK85/x/KK4l4IrQ9fwAAAABJRU5ErkJggg==" /><!-- --></p>
<p>The function also returns the average number of clones across the collection of trees for each patient. A clone is only counted if at least one gene is assigned to it.</p>
<p>Finally, the pairs of interest can also be visualized in a heatmap that shows in the respective patient in which clone the gene is mutated. The function <code>heatmap_clones_gene_pat</code> makes use of functionalities from the R packages <code>ggpubr</code> [<a href="https://CRAN.R-project.org/package=ggpubr">14</a>], and <code>reshape2</code> [<a href="http://www.jstatsoft.org/v21/i12/">15</a>].</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># extract the patient id's in which the pairs are mutated</span>
pairs_of_interest &lt;-<span class="st"> </span>sig_pairs <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(qval <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.02</span>)
pat_ids_of_interest &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">unlist</span>(<span class="kw">strsplit</span>(<span class="kw">as.character</span>(pairs_of_interest<span class="op">$</span>mutated_in), <span class="st">&quot;;&quot;</span>)))

<span class="co"># from the gene-to-clone tibble extract the data for one tree</span>
this_tree_id &lt;-<span class="st"> </span><span class="dv">1</span>
clone_tbl_all_pats_tree1 &lt;-<span class="st"> </span>clone_tbl_all_pats_all_trees <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">                   </span><span class="kw">filter</span>(tree_id <span class="op">==</span><span class="st"> </span>this_tree_id, patient_id <span class="op">%in%</span><span class="st"> </span>pat_ids_of_interest) <span class="op">%&gt;%</span><span class="st"> </span>
<span class="st">                   </span><span class="kw">select</span>(<span class="op">-</span>tree_id)

<span class="co"># plot the heatmap of clones and genes for the patients and genes of interest</span>
<span class="kw">heatmap_clones_gene_pat</span>(pairs_of_interest, clone_tbl_all_pats_tree1, all_genes_tbl, <span class="dt">first_clone_is_N =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAwAAAAHgCAMAAAAlhPoXAAABm1BMVEUAAAAAADoAAGYAAIsAOjoAOmYAOpAAZpAAZrY6AAA6ADo6AGY6OgA6Ojo6OmY6ZmY6ZpA6ZrY6kNtNTU1NTW5NTY5Nbm5Nbo5NbqtNjshmAABmADpmOgBmOjpmZjpmkJBmkLZmkNtmtttmtv9uTU1uTY5ubk1ubm5ubo5ubqtujo5ujqtujshuq8huq+SOTU2OTW6Obk2Obm6Ojm6Ojo6OjquOq6uOq8iOq+SOyOSOyP+QOgCQZjqQZmaQkDqQkLaQtpCQttuQ29uQ2/+rbk2rbm6rjm6rjo6rq46ryMiryOSr5P+t2Oa2ZgC2Zjq2ZpC2kGa2tra2ttu225C229u22/+2/7a2/9u2///Ijk3Ijm7Ijo7Iq27Iq47Iq6vIyKvIyMjIyOTI5OTI5P/I///bkDrbkGbbkJDbtmbbtpDb25Db27bb29vb2//b/7bb/9vb///kq27kyI7kyKvkyMjkyOTk5Kvk5Mjk5OTk5P/k///r6+v/tmb/yI7/25D/27b/29v/5Kv/5Mj/5OT//7b//8j//9v//+T////DT9vsAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAZR0lEQVR4nO2djX/cRlrH1+FS7KNX2mhTv5BCr12a0Lg2CbkmlwSOEl96JKTcAYU4kF7xS+kLR20H2tS1k7Mbe/fPZmYk7WptP/siSxpJz/f7+cRe72p2x7/H35mRZu00OgCKafjuAIBPEABUgwCgGgQA1SAAqAYBQDUIAKpBAFBNQoCdRuPMwxxeYr0R8srHRx44/MeHncNbjZc2xKb2iPjmR43GxJsbiSeczKGv/ohTmnj1aExZPfPp8n9+2/YtPvTBoFaVIiHAAxPS2zm8RFyAxsSdvvufXzgzpADuiBBzXPeH/kGdBch+FMok/2dTib7Zwah+AryYy+mHqlfa/tAeDC114oj1xsQHnc+coaaj53Lopmd6KWX93WWSvxl03rL52759ceHYc1WXngA7jYk/zWUNtB4+6/M5NwS59F65E43jZ/4tHIHcAue1DXf0xN+ZL852jwjHnFvWzfCjGYzymKg8E6Vkv7vJXJ75tPmbw8zQM+mEmpiqoQBmWfflVB5ja1zadVuAne5snCyAm31M6g8Ta+E7yQJEo75bepqn+MkJK9qKE6f0ReY1yCL/fgHO/svAM4dK0RXA/ogNPiNKS1SAr+bMZ/MKf7hhR7lz0QQbvqS5+XF0rynA2Y/tx3PJKTh8LLzjwYkr2qojLVSyeObT52/veMt+MHPvZ69t5POD4oWuADv2J2o9jx+ro6X96tcvu5ONXgGi8d1VKgw5XAicKMDh7SlToS/qMwmH9E5V38rrmU+Rf6fzkevbT8Mv6iiAW108y2MN1HcV4vB29xJOrwDRJYZGz8Fwsj15BoiftV5TQAEnwafIP74KF10HraEA3Qyy/8aiArgL3DbHs7/4aq6/ADuNYQU4LsBO7QQIv7Xnc1lficgif3PYGTftTsZf1k2A3gCU+RWW9URFwx/kF3PHRqC3e0efVIDkSXBIXQXIfmrLIv8jA1D9BIhmuGhyzJZkAXbiFea5I2vQyd7RwghkEneXQZ+9MXVkMqgF3RngVp4CpM2/9gLEV5/tVJf1z9WREeilDVPj8HKaiTq+CmFO/czkb24mC5AYDNftEW4jzLSY+KDzeeaXyz2TOAfI+Gcri/xfuE2E7pWH+gmwHi991rNfAyUL0J1pJt1odOQ6tI07WQB3xMNkQ3NnfLpSlwpE9ATIemmXSf47fX2rnQC9gf9Z9pcXkwVwVyHOfuDusruP8ZbK89vmp9q9DSxZAHdEtOF1+NFUI3oz3PM34m3LGtEVIPMdvmzyt6E34jfq1U4AAKUgAKgGAUA1CACqQQBQDQKAahAAVIMAoJoSCPB9we2gH935I4B6dOePAOrRnT8CqEd3/gigHt35I4B6dOePAOrRnX9BAjQKppjvqjqQv0QxXS06/ypVoAjIX6QoAf4gFY3GP6SiQgUoBPIXQQANkL8IAmiA/EUQQAPkL4IAGiB/EQTQAPmLIIAGyF8EATRA/iIIoAHyF0EADZC/CAJogPxFEEAD5C+CABogfxEE0AD5iyCABshfBAE0QP4iCKAB8hdBAA2QvwgCaID8RRBAA+QvggAaIH8RBNAA+YsggAbIXwQBNED+IgigAfIXQQANkL8IAmiA/EUQQAPkL4IAGiB/EQTQAPmLIIAGyF8EATRA/iIIoAHyF0EADZC/CAJogPxFhvf0ydUgCC6tmlvtpZnt0Z51N7jZ/yoUIDXknyvDetpeCkIWxijAXosCZAT558ywni4H0ysm+q1W8/7IBbBFowDZQP45M6Snu0GU+W4wO3IBNpu/ogDZQP55M6Sny2bgcbQ/WYkKsH+3FQSXV92M/M2VoHnDPrx/Nwia1+Na3WQNmhHknzeDe9o/5rivzPrScn6tuzxd6MR3ztqjDhZnOQnLCPLPncE9tWH2cAVYDua3O+1HbkYOZrc7mzb25eA9d6eNfdnUJlmA7w3FF8C+aj6JFQn55864AhwsujGpvXR+zf6zh8xs77Vmw8fNp02bPSNQNpB/7oy7BNprLbjbm8374YPhndG1uvhxCpAN5J87o54E21vDC3B+bTO6FXQbulehAOkg/7wZ0tO9VjQE2RvHpuBeARbiBhQgU8g/b0bYCLtnUn5s9xaPnYTFBWgvNW+Ym1uteMXKFJwV5J8zY78VIppuuxuTR+4MoQBZQf45M7Sn7a2L9s1YK534lMztucx/10kWwN3pdmdCKEBmkH++8HZoDZC/CAJogPxFEEAD5C+CABogfxEE0AD5iyCABshfBAE0QP4iCKAB8hdBAA2QvwgCaID8RRBAA+QvggAaIH8RBNAA+YsggAbIXwQBNED+IgigAfIXQQANkL8IAmiA/EUQQAPkL4IAGiB/EQTQAPmLIIAGyF8EATRA/iIIoAHyF0EADZC/CAJogPxFEEAD5C+CABogfxEE0AD5iyCABshfBAE0QP4iCKAB8hcpSoCiKeTbqgzkL4IAGiB/kaIESDuV1n4KLoT0OaZdOvn+jkcGATSAACIIoAEEEEEADSCACAJoAAFEEEADCCCCABpAABEE0AACiCCABhBABAE0gAAiCKABBBBBAA0ggAgCaAABRBBAAwggggAaQAARBNAAAogggAYQQAQBNIAAIgigAQQQQQANIIAIAmgAAUQQQAMIIIIAGkAAEQTQAAKIIIAGEEAEATSAACIIoAEEEEEADSCACAJoAAFEEEADCCCCABpAABEE0AACiCCABhBABAE0gAAiCKABBBAZ3tMnV4MguLRqbrWXZraHP2P7bhA0r/cfiADp8Zu/egHaS0HIwogFOLjiDp/tfxUESInv/NULsBxMr5jot1rN+6MVYDOY3+7sXzGHJ18FAVLiO3/tAuwGUea7ZkwZpQDRMZvBzb5XQYB0eM9fuwDL8UjS/mQlCnf/bisILq+6rL+5EjRv2If3+xee3yyeX+t7FQRIh/f8lQvQP+a4r/Zabo1pAo6Xp2Z1Gt0ZLTw3g2Bmtf9VECAV/vNXLsDBYvJkyhVg2a4x24/cjBzMbpu0Z+1K9T13ZzjvLl8Mgl7hvjcUL4B91VwCKxT/+acVoDL5jyvAwaKLtr10fs3+s4fMbO+1ZsPHu0dv9l+GYAZIh//8lc8Ax6fgvdaCu73ZvB8+GN4ZXavrHh0Wp/cqCJAK//krF6B3EmZvDS9AL/VlBMgC7/lrF2CvFQ0q9saxKbhXgIW4QfT4wSICZIH3/LULYDdi7pmUH7fMCdaxk7C4AO2l5g1zc6vlzsfsRsxSsJB8FgRIi+/81QtwbCs+mm67G5NH7jRjT/9qNHwVBEiJ7/zVC9Bpb120b8Za6cSnZG7PZf67TrIA7k63O8Ob4TLGc/4IkM2rIIBXEEAEATSAACIaBXjwyi823I3D/3rjXN8jv+8c/fLw1ksbo32Px9uWh3IJUKr8VQrQaLztbuw0Gn0F+OyPHh79cuQCnNC2PJRMgDLlr1KAiZcnwxs/muorwIMzDwd8OZjTtM2dkglQpvxVCnDmb1w8L+b+zH8BCqFkApQpf50C/POUnYN3zOdz0SRrPn59q9FoTHY6n7/caEy8uXHovgwffX57qtF49WN32JcXGhM/jZ/q8CPzwCsfdw7ltu5jfJwnyiZAifLXKcC/ztmR58FLX55QgPWG41wixGdT7q4zD8OgG72l6wP31cSdQ7Ft+PzxcflHfSJlE6BE+esU4OEDE8qLuXPPkgXYcPPmi7kzZqB4Nhd9GcX3mh1ETMIm2ckNk/Nk+Ewv7GGdz209pLbuY+I4L5RNgBLlr1SAHTMYmH8nFMDw37/565cbiRDD+MzNMw/tvzj3jr0x8epvfhc/6Ult4wJ0j/NC6QQoT/5KBTCjj5mBN04qQDTfJkJ8Fp2qrduptntweJebWd/c6Ehto4N7x3mhdAKUJ3+lAphU/sfU4MQpuPHHb/7TV3OjFaDz1e1E4ie0jQ/uHueF0glQnvyVCmDC/Nsf34kK0J1V7QM7bpPmRTLEvin4SAEMv//igpnQT2zbN1+Hx3mhdAKUJ3+tAjyb+on7aE+fGrYKbhSxK1N7lvX8QvzlsZOwvgI8m5o0K8vnt2wBjrftPXPiOC+UT4DS5K9VAHs5oRMWYMdOjmcvhCvFSTONxtfc1o9chovrkRiBwstr9qlOaJt45t5xXiifAKXJX6sAJpK3owJ0PptqvPq1zfSFHTzMCNL40Vt2BLFffh1txJj4X/td51gBDn89ZY82t05ou9F75t5xXiifAKXJX6MA+iiXAKUCATSAACIIoAEEEEEADSCAiEIBhDaFBOGJUglQrvwRwHMBCgEBRBDAcwEKAQFEEMBzAQoBAUQQwHMBCgEBRBDAcwEKAQFEEMBzAQoBAUQQoK8AD/y9Zz9PKiNA8fkjQLIA6y9tHN6aLCSSQqmKAB7yR4BEAV7MvW3/WEeZ/qJPNlREAB/5I0CiAM9+fCeqQs2oiAA+8keARAF2wl+hQ4BejoUK4CP/ogQomgGdGVoAX3+9Jz/IXwQBPI9AhUD+ImVfAtV+DVoIFVmCcg5wPMhCBeAq0LH807Yb0Jly5Y8AiQLYP1bGPkBf/mnbDehMufJHgGQB7E7kZCGJFEtVBPCQPwL0FaCmVEaA4kEAzwUoBAQQQQDPBSgEBBBBAM8FKAQEEFEogEJKJUC5UCiA0KaQIDxRKgHKlT8CIMCg/NO2G9CZcuWPAAgwKP+07QZ0plz5IwACDMo/bbsBnSlX/giAAIPyT9tuQGfKlT8CIMCg/NO2G9CZcuWPAAgwKP+07QZ0plz5I0B/AV78iaf/xy5XqiNA4fkjQH8BHvj6jxxzpToCFJ4/AiQL8GKugQB9+adtN6Az5cofAZIF2Jl0v5VXO6oigIf8ESBZgOjXUmtHVQTwkD8CIMCg/NO2G9CZcuWPAAgwKP+07QZ0plz5IwACDMo/bbsBnSlX/giAAIPyT9tuQGfKlT8CIMCg/NO2G9CZcuWPAAgwKP+07QZ0plz5I0B/AepJdQQoHARAgEH5p203oDPlyh8BEGBQ/mnbDehMufJHAAQYlH/adgM6U678EQABBuWftt2AzpQrf4UCKKRUApQLBNAAAogggAYQQGR4T59cDYLg0qq51V6a2R7+jPvm+OZ8/4EIkB6/+adtl08WOTCsp+2lIGRhxALstdzh02t9r4IAKfGdf9p2uQWSNcN6uhxMr5jot1rN+yMVwBTs52YUWrIFS7wKAqTEd/5p2+WXSMYM6eluEGW+G8yOVICDxdnEp+6rIEA6vOeftl0uaeTBkJ4um4HH0f5kJSrA/l0zyV5edTPyN1eC5g378P5ds/C83isPAmSD9/zTtiswo9MxuKf9Y477Klpjnl/rLk8XugvPXui7LIGywH/+adsVFtFpGdzT/oHEFWA5mN/utB+5GTmY3e5s2tiXg/fcnTejI/davcJ9byheAPuqOcRVMP7zT9uuMvmPK8DBoou2vXR+zf6zh8xs77Vmw8ejo/da8cwdvwozQCr855+2XUEBnZ5xl0B7rXBu3WzeDx8M74yu1YVHbwbTq0deBQFS4T//tO0Ki+i0jHoSbG8NL4AdkszMfOxiBQKkxHv+adsVmtJpGNLT7mLS3jg2BfcK0Dvlah+5BB2+CgKkw3v+adsVFdCpGWEj7J5J9XHLnGAdOwmLC9Beat4wN7da7nzs5vFnQYC0+M4/bbvik0rJ2G+FiKbb7sbkkTv7ZuPeqyBASnznn7adt8DGZWhP21sX7ZuxVjrxKZnbc5n/rpMsgLvT7c7sBgiQKZ7zT9vOV1xjw9uhNYAAIgigAQQQQQANIIAIAmgAAUQQQAMIIIIAGkAAEQTQAAKIIIAGEEAEATSAACIIoAEEEEEADSCACAJoAAFEEEADCCCCABpAABEE0AACiCCABhBABAE0gAAiCKABBBBBAA0ggAgCaAABRBBAAwggggAaQAARBNAAAogggAYQQAQBNIAAIgigAQQQQQANIIAIAmgAAUQQQAMIIIIAGkAAEQTQAAKIIIAGEECkKAGKppBvqzKQvwgCaID8Rcq+BKr9FFwI5C+CABogfxEE0AD5iyCABshfBAE0QP4iCKAB8hdBAA2QvwgCaID8RRBAA+QvggAaIH8RBNAA+YsggAbIXwQBNED+IgigAfIXQQANkL8IAmiA/EUQQAPkL4IAGiB/EQTQAPmLIIAGyF8EATRA/iIIoAHyF0EADZC/CAJogPxFEEAD5C+CABogfxEE0AD5iyCABshfBAE0QP4iCKAB8hdBAA2QvwgCaID8RRBAA+QvggAaIH8RBNAA+YsM7+mTq0EQXFo1t9pLM9sjPenB4s0jr0IBUkP+uTKsp+2lIGRh9AIcXAkoQEaQf84M6+lyML1iot9qNe+PWoAnrYACZAX558yQnu4GUea7wexoBWj/Mpj+FQXICPLPmyE9XTYDj6P9yUpUgP27ZoS5vOpm5G+uBM0b9uH9u0HQvG7Lc/CX17d3KUBGkH/eDO5p/5jjvtpruSXp+bXu8tSsTqM7Z6MDKUBGkH/uDO7pweJs4itXgOVgfrvTfuRm5GB2u7NpY18O3nN3Rrn3FeB7Q/EFsK+afVpFQ/65M64AB4tuTGovnV+z/+whM9t7rdnw8ehoRqCMIP/cGXcJtNdacLc3m/fDB8M7o2t13TM2CpAJ5J87o54E21vDC2CHpA4FyA7yz5shPd1rRYOKvXFsCu4VYKGvFQXICvLPmxE2wu6ZlB+3TKbHTsLiArSXmjfMza0Wa9CsIf+cGfutENF0292YPHKngwJkBvnnzNCetrcu2jdjrXTiUzK35zL/XSdZAHen251xUIDsIP984e3QGiB/EQTQAPmLIIAGyF8EATRA/iIIoAHyF0EADZC/CAJogPxFEEAD5C+CABogfxEE0AD5iyCABshfBAE0QP4iCKAB8hdBAA2QvwgCaID8RRBAA+QvggAaIH8RBNAA+YsggAbIXwQBNED+IgigAfIXQQANkL8IAmiA/EUQQAPkL4IAGiB/EQTQAPmLIIAGyF8EATRA/iIIoAHyF0EADZC/CAJogPxFEEAD5C9SlABFU8i3VRnIX6SgnlIAv5C/RAl6mvZ/FK/G/0RefnTnjwDq0Z0/AqhHd/4IoB7d+SOAenTnjwDq0Z0/AqhHd/4lEADAHwgAqkEAUA0CgGoQAFSDAKAaBADVIACoxrcA7f/b7nSefHjtZ6vjt/02XTNIoD5/zwJsBufX2kuBZX6MZlvvBPPbm67ZQm590wD5+xVgM2jeMx+mVzv7i2NEuRsE7wTvtmZWO09awc0c+1d3yN+vAO2l5v3wgw11ZnvkZib1XTN2jdcMjkL+ngU4WDQhug+d+NMYzVzyYzSDo5B/OWaAVAVo3/2LWhTAJ+Tv/xzADCObbhm5GcyO3Gy5t14dpxkchfw9C2BWk69f//RR8+dP/yMYYyQ5WIyObv8ycAtYSAf5+74M2n7UCkIuj3NFuf3oz10BDhanK38h2ivk73sjrNP54X8//fTTFd+90Ivy/P0LAOCRKgvw5Nq196s+A1eZWuRfRQHCS2/RDn7Vt+IrSK3yr64A8Q5+xbfiK0it8ve8ExwkGHMjZuwdfDgK+fueAZ60gnevxfxs1CRT7uDDMcjf9xJor5ViAIlGoHoUwC/k7/scYDfFWZSZud99/9Otlm1Z+a14z6jP37cAneXxR5D2hxejRWsNtuJ9oz1/7wKkpP3tb69dWqvBVnxFqU3+VRUAIBMQAFTjW4D2EzuHtpfmvxurmf07BnrfwJUh6vP3fhnUXYXYXwyaY+woRrvwFd+CKQPk71cAk/97YYpb41xO2AxmVjv7S5V/H4p3yN+zAOl+ty7ahU+1iQNJyL8EvxQfMUacddmF9w75e38zXC/Ccf8qQV0K4BPyRwDVkL/3JVDv3eRjvK+2VgXwCfn7Pgnupd4e45qCSX7l6dOn34afxruADUnI3//fBZpxbybZXxrjqnLK3+OAY5C/740wu6Xy+qWrrbF2Vdr2D3l0+c8aXIvzBvn7fitE54lNv3m5HvvqFUR7/t4FSMMPFV93Vp065V9FAepx+aG61Cp/vwL88DRirHVkrQrgFfL3vRHWu5gwP3oNalUAn5C/742wD+M/yfFXrWB65FBrVQCfkL/vJVCP8TZimr2/ZjP6n7OBAajNvywCjPduxBptxJQFrfmXRoDx34wFWaI1fwQAh9b8SyPA+L+QAVmiNf/SCLA8+q/k1aoAZUFr/iXZCPvt1TF+Kbv9YcWvPJQG8i/PRth0xf/GZBUhf98bYb331dbo7VXVgfx9L4EAPIMAoJqSnAQ/rf4vl1YR8vf+VyHqs6deQcjf9wywGTRr866qKkL+vpdAd1vBpXteu6Aa8vd/Evzt3cDUQOXgUwq05+9dAMOTK0EwrbgGvlGdfxkEMGdjj9+pyf+3UE0U518OATo/mArovApRDvTmXwYBfnhs/zbT+1X/Dzcri+r8vQvgZt/Xr+vchSkB2vP3/Mdxbfpqz7/8Q/7eN8Km/175VrxXyJ+3QmiG/H0L8O/XrmnfivcJ+fteAgF4BgFANQgAqkEAUA0CgGoQAFSDAKAaBADVIACoBgFANQgAqkEAUA0CgGoQAFSDAKAaBADVIACoBgFANQgAqkEAUA0CgGoQAFSDAKAaBADVIACoBgFANQgAqkEAUA0CgGr+H7m6Xv+/db+IAAAAAElFTkSuQmCC" /><!-- --></p>
<hr />
</div>
<div id="pathway-level-analysis" class="section level2">
<h2>Pathway level analysis</h2>
<p>Especially if you have whole-exome sequencing data, the GeneAccord analysis can also be done on the pathway level. In that case, the mutated genes that are assigned to clones would be mapped to pathways. As an example, we demonstrate here how to convert a tibble that contains the gene-to-clone assignments into a tibble that contains pathway-to-clone assignments using the Reactome pathway data base [<a href="https://reactome.org/">16</a>].</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">data</span>(<span class="st">&quot;ensg_reactome_path_map&quot;</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">clone_tbl_pat01_tree1 &lt;-<span class="st"> </span>clone_tbl_all_pats_all_trees <span class="op">%&gt;%</span>
<span class="st">                            </span><span class="kw">filter</span>(patient_id <span class="op">==</span><span class="st"> &quot;01&quot;</span>) <span class="op">%&gt;%</span>
<span class="st">                            </span><span class="kw">filter</span>(tree_id <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) <span class="op">%&gt;%</span>
<span class="st">                            </span><span class="kw">select</span>(<span class="op">-</span>tree_id)

clone_tbl_pat01_tree1_pw &lt;-<span class="st"> </span><span class="kw">convert_ensembl_to_reactome_pw_tbl</span>(clone_tbl_pat01_tree1, ensg_reactome_path_map)
<span class="co">#&gt; Found 29 different Ensembl gene id's in the provided alteration tibble.</span>
<span class="co">#&gt; 9 could not be mapped to a reactome pathway.</span>
<span class="co">#&gt; The remaining 20 different Ensembl gene id's were mapped to 110 different reactome pathways.</span>
<span class="co">#&gt; There are 110 clone-pathway associations in the converted tibble.</span>


clone_tbl_pat01_tree1_pw
<span class="co">#&gt; # A tibble: 119 x 10</span>
<span class="co">#&gt;    file_name  patient_id altered_entity clone1 clone2 clone3 clone4 clone5</span>
<span class="co">#&gt;    &lt;fct&gt;      &lt;fct&gt;      &lt;fct&gt;           &lt;int&gt;  &lt;int&gt;  &lt;int&gt;  &lt;int&gt;  &lt;int&gt;</span>
<span class="co">#&gt;  1 cloe_seed~ 01         ABC-family pr~      0      1      1      1      0</span>
<span class="co">#&gt;  2 cloe_seed~ 01         Activation of~      0      1      1      1      0</span>
<span class="co">#&gt;  3 cloe_seed~ 01         Activation of~      0      1      1      1      0</span>
<span class="co">#&gt;  4 cloe_seed~ 01         Antigen proce~      0      1      1      1      0</span>
<span class="co">#&gt;  5 cloe_seed~ 01         APC/C:Cdc20 m~      0      1      1      1      0</span>
<span class="co">#&gt;  6 cloe_seed~ 01         APC/C:Cdh1 me~      0      1      1      1      0</span>
<span class="co">#&gt;  7 cloe_seed~ 01         Assembly of c~      0      1      1      1      0</span>
<span class="co">#&gt;  8 cloe_seed~ 01         Asymmetric lo~      0      1      1      1      0</span>
<span class="co">#&gt;  9 cloe_seed~ 01         AUF1 (hnRNP D~      0      1      1      1      0</span>
<span class="co">#&gt; 10 cloe_seed~ 01         Autodegradati~      0      1      1      1      0</span>
<span class="co">#&gt; # ... with 109 more rows, and 2 more variables: clone6 &lt;int&gt;, clone7 &lt;int&gt;</span></code></pre></div>
<hr />
</div>
</div>
<div id="other-alternative" class="section level1">
<h1>Other alternative</h1>
<p>The GeneAccord framework also offers the option of testing two-sided, meaning that you test pairs for significance that tend to be in different clones and also pairs that tend to be in the same clones more often than expected. The number of pairs to test with option <code>alternative = 'two.sided'</code> increases a lot, therefore, GeneAccord includes the additional option of testing only specific pre-defined genes of interest.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># this tells GeneAccord that both cases should be tested: </span>
<span class="co"># pairs that tend to co-occur together in the same clone more often than expected,</span>
<span class="co"># and pairs that tend to be clonally exclusive.</span>
alternative &lt;-<span class="st"> &quot;two.sided&quot;</span>

<span class="co"># to limit the number of tests done, the genes that should be</span>
<span class="co"># tested can be pre-specified</span>
<span class="co"># here, the genes VHL, PXDN</span>
genes_of_interest &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;ENSG00000134086&quot;</span>, <span class="st">&quot;ENSG00000130508&quot;</span>)

<span class="co"># this tells GeneAccord that we only want to test pairs where both gene A AND gene B</span>
<span class="co"># are from the 'genes of interest', so we're only testing the pair {VHL, PXDN}</span>
AND_OR &lt;-<span class="st"> &quot;AND&quot;</span>

<span class="co"># run GeneAccord with these options</span>
res_pairs_ts &lt;-<span class="st"> </span><span class="kw">suppressMessages</span>(<span class="kw">GeneAccord</span>(clone_tbl_all_pats_all_trees, avg_rates_m, ecdf_list, alternative, genes_of_interest, AND_OR))

res_pairs_ts
<span class="co">#&gt; # A tibble: 1 x 7</span>
<span class="co">#&gt;   entity_A   entity_B  num_patients   pval mle_delta test_statistic   qval</span>
<span class="co">#&gt;   &lt;fct&gt;      &lt;fct&gt;            &lt;int&gt;  &lt;dbl&gt;     &lt;dbl&gt;          &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt; 1 ENSG00000~ ENSG0000~            2 0.0589     -27.4          0.762 0.0589</span></code></pre></div>
<p>The parameter <code>AND_OR</code> can also be set to <code>OR</code>, meaning that all pairs will be tested where at least one of the two genes is in the set of genes of interest.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">AND_OR &lt;-<span class="st"> &quot;OR&quot;</span>

<span class="co"># Then, to limit the number of tests performed, it is recommended to only specify a small</span>
<span class="co"># number of genes</span>
<span class="co"># here, only SART3</span>
genes_of_interest &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;ENSG00000075856&quot;</span>)

<span class="co"># run GeneAccord testing all pairs that include the gene SART3</span>
res_pairs_ts &lt;-<span class="st"> </span><span class="kw">suppressMessages</span>(<span class="kw">GeneAccord</span>(clone_tbl_all_pats_all_trees, avg_rates_m, ecdf_list, alternative, genes_of_interest, AND_OR))

res_pairs_ts
<span class="co">#&gt; # A tibble: 2 x 7</span>
<span class="co">#&gt;   entity_A   entity_B  num_patients   pval mle_delta test_statistic   qval</span>
<span class="co">#&gt;   &lt;fct&gt;      &lt;fct&gt;            &lt;int&gt;  &lt;dbl&gt;     &lt;dbl&gt;          &lt;dbl&gt;  &lt;dbl&gt;</span>
<span class="co">#&gt; 1 ENSG00000~ ENSG0000~            2 0.0665     -24.7          0.662 0.0665</span>
<span class="co">#&gt; 2 ENSG00000~ ENSG0000~            2 0.0596    -184.           0.749 0.0665</span></code></pre></div>
<hr />
</div>
<div id="session-information" class="section level1">
<h1>Session information</h1>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">sessionInfo</span>()
<span class="co">#&gt; R version 3.4.4 (2018-03-15)</span>
<span class="co">#&gt; Platform: x86_64-w64-mingw32/x64 (64-bit)</span>
<span class="co">#&gt; Running under: Windows &gt;= 8 x64 (build 9200)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt; [1] LC_COLLATE=German_Switzerland.1252  LC_CTYPE=German_Switzerland.1252   </span>
<span class="co">#&gt; [3] LC_MONETARY=German_Switzerland.1252 LC_NUMERIC=C                       </span>
<span class="co">#&gt; [5] LC_TIME=German_Switzerland.1252    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt; [1] dplyr_0.7.4       magrittr_1.5      bindrcpp_0.2      GeneAccord_0.99.0</span>
<span class="co">#&gt; [5] digest_0.6.15    </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;  [1] Biobase_2.38.0       httr_1.3.1           bit64_0.9-7         </span>
<span class="co">#&gt;  [4] R.utils_2.6.0        gtools_3.5.0         assertthat_0.2.0    </span>
<span class="co">#&gt;  [7] highr_0.6            stats4_3.4.4         blob_1.1.0          </span>
<span class="co">#&gt; [10] yaml_2.1.18          progress_1.1.2       pillar_1.2.1        </span>
<span class="co">#&gt; [13] RSQLite_2.0          backports_1.1.2      lattice_0.20-35     </span>
<span class="co">#&gt; [16] glue_1.2.0           RColorBrewer_1.1-2   colorspace_1.3-2    </span>
<span class="co">#&gt; [19] sandwich_2.4-0       cowplot_0.9.2        htmltools_0.3.6     </span>
<span class="co">#&gt; [22] R.oo_1.21.0          plyr_1.8.4           XML_3.98-1.10       </span>
<span class="co">#&gt; [25] pkgconfig_2.0.1      devtools_1.13.5      biomaRt_2.34.2      </span>
<span class="co">#&gt; [28] purrr_0.2.4          scales_0.5.0         git2r_0.21.0        </span>
<span class="co">#&gt; [31] tibble_1.4.2         IRanges_2.12.0       ggplot2_2.2.1       </span>
<span class="co">#&gt; [34] ggpubr_0.1.6         withr_2.1.1          maxLik_1.3-4        </span>
<span class="co">#&gt; [37] BiocGenerics_0.24.0  lazyeval_0.2.1       cli_1.0.0           </span>
<span class="co">#&gt; [40] crayon_1.3.4         memoise_1.1.0        evaluate_0.10.1     </span>
<span class="co">#&gt; [43] R.methodsS3_1.7.1    xml2_1.2.0           BiocInstaller_1.28.0</span>
<span class="co">#&gt; [46] tools_3.4.4          prettyunits_1.0.2    stringr_1.3.0       </span>
<span class="co">#&gt; [49] S4Vectors_0.16.0     munsell_0.4.3        AnnotationDbi_1.40.0</span>
<span class="co">#&gt; [52] compiler_3.4.4       caTools_1.17.1       rlang_0.2.0         </span>
<span class="co">#&gt; [55] grid_3.4.4           RCurl_1.95-4.10      rstudioapi_0.7      </span>
<span class="co">#&gt; [58] miscTools_0.6-22     labeling_0.3         bitops_1.0-6        </span>
<span class="co">#&gt; [61] rmarkdown_1.9        testthat_2.0.0       gtable_0.2.0        </span>
<span class="co">#&gt; [64] DBI_0.8              roxygen2_6.0.1       reshape2_1.4.3      </span>
<span class="co">#&gt; [67] R6_2.2.2             zoo_1.8-1            knitr_1.20          </span>
<span class="co">#&gt; [70] bit_1.1-12           utf8_1.1.3           bindr_0.1.1         </span>
<span class="co">#&gt; [73] commonmark_1.4       rprojroot_1.3-2      desc_1.1.1          </span>
<span class="co">#&gt; [76] stringi_1.1.7        parallel_3.4.4       Rcpp_0.12.15</span></code></pre></div>
<hr />
<div id="references" class="section level2">
<h2>References:</h2>
<p>[1] Moore AL, Kuipers J, Singer J, Burcklen E, Schraml P, Beisel C, Moch H, Beerenwinkel N. Intra-tumor heterogeneity and clonal exclusivity in renal cell carcinoma. (2018). bioRxiv number. doi: doi</p>
<p>[2] Benjamini Y, Hochberg Y. Controlling the false discovery rate: a practical and 26 powerful approach to multiple testing. Journal of the royal statistical society Series B 27 (Methodological). 1995:289-300 %@ 0035-9246.</p>
<p>[3] Beerenwinkel N, Schwarz RF, Gerstung M, Markowetz F. Cancer evolution: mathematical models and computational inference. Syst Biol. 2015;64(1):e1-25.</p>
<p>[4] Marass F, Mouliere F, Yuan K, Rosenfeld N, Markowetz F. A phylogenetic latent feature model for clonal deconvolution. The Annals of Applied Statistics. 2016;10(4):2377-404 %@ 1932-6157.</p>
<p>[5] Kirill Mueller and Hadley Wickham (2018). tibble: Simple Data Frames. R package version 1.4.2. <a href="https://CRAN.R-project.org/package=tibble" class="uri">https://CRAN.R-project.org/package=tibble</a></p>
<p>[6] Jarek Tuszynski (2014). caTools: Tools: moving window statistics, GIF, Base64, ROC AUC, etc.. R package version 1.17.1. <a href="https://CRAN.R-project.org/package=caTools" class="uri">https://CRAN.R-project.org/package=caTools</a></p>
<p>[7] Gregory R. Warnes, Ben Bolker and Thomas Lumley (2015). gtools: Various R Programming Tools. R package version 3.5.0. <a href="https://CRAN.R-project.org/package=gtools" class="uri">https://CRAN.R-project.org/package=gtools</a></p>
<p>[8] Arne Henningsen and Ott Toomet (2011). maxLik: A package for maximum likelihood estimation in R. Computational Statistics 26(3), 443-458. DOI 10.1007/s00180-010-0217-1.</p>
<p>[9] Stefan Milton Bache and Hadley Wickham (2014). magrittr: A Forward-Pipe Operator for R. R package version 1.5. <a href="https://CRAN.R-project.org/package=magrittr" class="uri">https://CRAN.R-project.org/package=magrittr</a></p>
<p>[10] Durinck S, Moreau Y, Kasprzyk A, Davis S, De Moor B, Brazma A and Huber W (2005). “BioMart and Bioconductor: a powerful link between biological databases and microarray data analysis.” Bioinformatics, 21, pp. 3439-3440.</p>
<p>[11] Durinck S, Spellman P, Birney E and Huber W (2009). “Mapping identifiers for the integration of genomic datasets with the R/Bioconductor package biomaRt.” Nature Protocols, 4, pp. 1184-1191.</p>
<p>[12] Hadley Wickham, Romain Francois, Lionel Henry and Kirill Mueller (2017). dplyr: A Grammar of Data Manipulation. R package version 0.7.4. <a href="https://CRAN.R-project.org/package=dplyr" class="uri">https://CRAN.R-project.org/package=dplyr</a></p>
<p>[13] H. Wickham. ggplot2: Elegant Graphics for Data Analysis. Springer-Verlag New York, 2009.</p>
<p>[14] Alboukadel Kassambara (2017). ggpubr: ‘ggplot2’ Based Publication Ready Plots. R package version 0.1.6. <a href="https://CRAN.R-project.org/package=ggpubr" class="uri">https://CRAN.R-project.org/package=ggpubr</a></p>
<p>[15] Hadley Wickham (2007). Reshaping Data with the reshape Package. Journal of Statistical Software, 21(12), 1-20. URL <a href="http://www.jstatsoft.org/v21/i12/" class="uri">http://www.jstatsoft.org/v21/i12/</a>.</p>
<p>[16] Matthews L, Gopinath G, Gillespie M, Caudy M, Croft D, de Bono B, et al. Reactome knowledgebase of human biological pathways and processes. Nucleic acids research. 2009;37(Database issue):D619-22.</p>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
